<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Reference Data QC: MUGA</title>

<script src="site_libs/header-attrs-2.16/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/spacelab.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<script src="site_libs/htmlwidgets-1.5.4/htmlwidgets.js"></script>
<link href="site_libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet" />
<script src="site_libs/datatables-binding-0.25/datatables.js"></script>
<link href="site_libs/dt-core-1.11.3/css/jquery.dataTables.min.css" rel="stylesheet" />
<link href="site_libs/dt-core-1.11.3/css/jquery.dataTables.extra.css" rel="stylesheet" />
<script src="site_libs/dt-core-1.11.3/js/jquery.dataTables.min.js"></script>
<link href="site_libs/nouislider-7.0.10/jquery.nouislider.min.css" rel="stylesheet" />
<script src="site_libs/nouislider-7.0.10/jquery.nouislider.min.js"></script>
<link href="site_libs/selectize-0.12.0/selectize.bootstrap3.css" rel="stylesheet" />
<script src="site_libs/selectize-0.12.0/selectize.min.js"></script>
<link href="site_libs/crosstalk-1.2.0/css/crosstalk.min.css" rel="stylesheet" />
<script src="site_libs/crosstalk-1.2.0/js/crosstalk.min.js"></script>
<script src="site_libs/plotly-binding-4.10.0/plotly.js"></script>
<script src="site_libs/typedarray-0.1/typedarray.min.js"></script>
<link href="site_libs/plotly-htmlwidgets-css-2.5.1/plotly-htmlwidgets.css" rel="stylesheet" />
<script src="site_libs/plotly-main-2.5.1/plotly-latest.min.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">JAX GeDI Reference Genotype Data Repository</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="MUGA_Reference_QC.html">MUGA Data QC Overview</a>
</li>
<li>
  <a href="MegaMUGA_Reference_QC.html">MegaMUGA QC Overview</a>
</li>
<li>
  <a href="GigaMUGA_Reference_QC.html">GigaMUGA QC Overview</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/sam-widmayer/MUGA_reference_data">
    <span class="fab fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Reference Data QC: MUGA</h1>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2022-11-09
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>MUGA_reference_data/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.0). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20221108code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20221108)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20221108code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20221108)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomsamwidmayerMUGAreferencedatatree48d211ef10e0fef9b875f26067898cb12e24671btargetblank48d211ea">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/sam-widmayer/MUGA_reference_data/tree/48d211ef10e0fef9b875f26067898cb12e24671b" target="_blank">48d211e</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomsamwidmayerMUGAreferencedatatree48d211ef10e0fef9b875f26067898cb12e24671btargetblank48d211ea"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/sam-widmayer/MUGA_reference_data/tree/48d211ef10e0fef9b875f26067898cb12e24671b" target="_blank">48d211e</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    data/.DS_Store
    Ignored:    output/.DS_Store

Untracked files:
    Untracked:  README.html
    Untracked:  data/GigaMUGA/FPGM_1-6_FinalReport.txt
    Untracked:  data/GigaMUGA/FPGM_1-6_FinalReport.zip
    Untracked:  data/GigaMUGA/GigaMUGA_control_genotypes.txt
    Untracked:  data/GigaMUGA/GigaMUGA_founder_consensus_genotypes/
    Untracked:  data/GigaMUGA/GigaMUGA_founder_sample_concordance/
    Untracked:  data/GigaMUGA/GigaMUGA_founder_sample_dendrogram_genos/
    Untracked:  data/GigaMUGA/GigaMUGA_founder_sample_genotypes/
    Untracked:  data/GigaMUGA/GigaMUGA_reference_genotypes/
    Untracked:  output/GigaMUGA/
    Untracked:  readme_diagram.png

Unstaged changes:
    Modified:   analysis/_site.yml
    Modified:   analysis/license.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/MUGA_Reference_QC.Rmd</code>) and
HTML (<code>docs/MUGA_Reference_QC.html</code>) files. If you’ve
configured a remote Git repository (see <code>?wflow_git_remote</code>),
click on the hyperlinks in the table below to view the files as they
were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/sam-widmayer/MUGA_reference_data/blob/d352284587586c6a8fe7e75630bd27ff5b9f5894/analysis/MUGA_Reference_QC.Rmd" target="_blank">d352284</a>
</td>
<td>
sam-widmayer
</td>
<td>
2022-11-09
</td>
<td>
update site theme and fold code snippets
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/sam-widmayer/MUGA_reference_data/d352284587586c6a8fe7e75630bd27ff5b9f5894/docs/MUGA_Reference_QC.html" target="_blank">d352284</a>
</td>
<td>
sam-widmayer
</td>
<td>
2022-11-09
</td>
<td>
update site theme and fold code snippets
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/sam-widmayer/MUGA_reference_data/blob/3501596652f350afae3250facfce0abffc9ee2f5/analysis/MUGA_Reference_QC.Rmd" target="_blank">3501596</a>
</td>
<td>
sam-widmayer
</td>
<td>
2022-11-09
</td>
<td>
change site theme and add index tables
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/sam-widmayer/MUGA_reference_data/3501596652f350afae3250facfce0abffc9ee2f5/docs/MUGA_Reference_QC.html" target="_blank">3501596</a>
</td>
<td>
sam-widmayer
</td>
<td>
2022-11-09
</td>
<td>
change site theme and add index tables
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/sam-widmayer/MUGA_reference_data/blob/107452926d1aa578d253488e5c91c3bc8b1cbfdf/analysis/MUGA_Reference_QC.Rmd" target="_blank">1074529</a>
</td>
<td>
sam-widmayer
</td>
<td>
2022-11-08
</td>
<td>
update markdown formatting for workflowr
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/sam-widmayer/MUGA_reference_data/blob/efe8849090933ad18a23b06fdfdef16e6c6676c4/analysis/MUGA_Reference_QC.Rmd" target="_blank">efe8849</a>
</td>
<td>
sam-widmayer
</td>
<td>
2022-11-08
</td>
<td>
initialize workflowr integration
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="muga-annotations" class="section level1">
<h1>MUGA Annotations</h1>
<div id="reading-in-reference-genotypes-and-metadata"
class="section level2">
<h2>Reading in reference genotypes and metadata</h2>
<p>First I read in the reference sample genotypes, as well as marker
annotations from an <a
href="https://github.com/kbroman/MUGAarrays">analysis</a> previously
conducted by Karl Broman, Dan Gatti, and Belinda Cornes.</p>
<pre class="r"><code>## Reading in marker annotations fro Broman, Gatti, &amp; Cornes analysis
pre_control_genotypes &lt;- data.table::fread(&quot;data/MUGA/vendu.dodata.csv&quot;, header = T)
samples &lt;- unique(colnames(pre_control_genotypes)[-c(1:3)])
control_genotypes &lt;- data.table::fread(&quot;data/MUGA/vendu.dodata.csv&quot;, header = T, skip = 1)


tidyMUGAgenos &lt;- function(smp){
    # print(smp)
    meta &lt;- control_genotypes[,1:3]
    geno_info &lt;- data.frame(control_genotypes)[,which(colnames(pre_control_genotypes) == smp)]
    if(ncol(geno_info &gt; 8)){
      # B6 sample is exactly repeated in source data
      geno_info_1 &lt;- geno_info[,1:8]
      tidy_geno_info &lt;- cbind(meta, geno_info_1) %&gt;%
        dplyr::mutate(sample = paste0(smp)) %&gt;%
        `colnames&lt;-`(c(&quot;snpID&quot;,&quot;Chromosome&quot;,&quot;Position&quot;,&quot;call&quot;,&quot;X&quot;,&quot;Y&quot;,&quot;gcScore&quot;,&quot;theta&quot;,&quot;r&quot;,&quot;xRaw&quot;,&quot;yRaw&quot;,&quot;sample&quot;))
      return(tidy_geno_info)
      
    } else {
      tidy_geno_info &lt;- cbind(meta, geno_info) %&gt;%
        dplyr::mutate(sample = smp) %&gt;%
        `colnames&lt;-`(c(&quot;snpID&quot;,&quot;Chromosome&quot;,&quot;Position&quot;,&quot;call&quot;,&quot;X&quot;,&quot;Y&quot;,&quot;gcScore&quot;,&quot;theta&quot;,&quot;r&quot;,&quot;xRaw&quot;,&quot;yRaw&quot;,&quot;sample&quot;))
      return(tidy_geno_info)  
    }
    
}
tidy_control_genotypes &lt;- purrr::map(samples, tidyMUGAgenos) %&gt;% 
  Reduce(rbind,.) %&gt;%
  dplyr::rename(marker = snpID,
                genotype = call)
muga_metadata &lt;- data.table::fread(&quot;data/MUGA/muga_uwisc_v2.csv&quot;, header = T)</code></pre>
</div>
<div id="marker-qc-searching-for-missing-genotype-calls"
class="section level2">
<h2>Marker QC: Searching for missing genotype calls</h2>
<p>We searched for probes where many mice are missing genotype
calls.</p>
<pre class="r"><code>## Calculating allele frequencies for each marker
control_allele_freqs &lt;- tidy_control_genotypes %&gt;%
  dplyr::select(sample, marker, genotype) %&gt;%
  dplyr::group_by(marker, genotype) %&gt;%
  dplyr::count() %&gt;%
  # Result: number of genotype calls for each marker across all samples
  # i.e.
#    marker            genotype     n
#    &lt;chr&gt;             &lt;chr&gt;    &lt;int&gt;
#  1 backupJAX00000484 A           12
#  2 backupJAX00000484 G           21
#  3 backupJAX00000484 H           22
#  4 backupJAX00003592 A           14
#  5 backupJAX00003592 G           18
#  6 backupJAX00003592 H           23
#  7 backupJAX00004293 C           16
#  8 backupJAX00004293 H           24
#  9 backupJAX00004293 T           15
# 10 backupJAX00005508 A            8
  dplyr::ungroup() %&gt;%
  dplyr::group_by(marker) %&gt;%
  dplyr::mutate(freq = round(n/sum(n), 3),
                genotype = as.factor(genotype)) %&gt;%
  # Result: allele frequency calls for each marker across all samples
  # i.e.
#    marker            genotype     n  freq
#    &lt;chr&gt;             &lt;fct&gt;    &lt;int&gt; &lt;dbl&gt;
#  1 backupJAX00000484 A           12 0.218
#  2 backupJAX00000484 G           21 0.382
#  3 backupJAX00000484 H           22 0.4  
#  4 backupJAX00003592 A           14 0.255
#  5 backupJAX00003592 G           18 0.327
#  6 backupJAX00003592 H           23 0.418
#  7 backupJAX00004293 C           16 0.291
#  8 backupJAX00004293 H           24 0.436
#  9 backupJAX00004293 T           15 0.273
# 10 backupJAX00005508 A            8 0.145
  dplyr::left_join(., muga_metadata)


## Filtering to markers with missing genotypes
no.calls &lt;- control_allele_freqs %&gt;%
  dplyr::ungroup() %&gt;%
  dplyr::filter(genotype == &quot;N&quot;) %&gt;%
  tidyr::pivot_wider(names_from = genotype, 
                     values_from = n) %&gt;%
  dplyr::select(marker, chr, bp_grcm39, freq) %&gt;%
  dplyr::mutate(chr = as.factor(chr))


## Identifying markers with missing genotypes at a frequency higher than the 95th percentile of &quot;N&quot; frequencies across all markers
cutoff &lt;- quantile(no.calls$freq, probs = seq(0,1,0.05))[[20]]
above.cutoff &lt;- no.calls %&gt;%
  dplyr::filter(freq &gt; cutoff)</code></pre>
<p>Of 7854 markers, 1832 failed to genotype at least one sample, and 92
markers failed to genotype at least 37.21% of samples.</p>
<div id="htmlwidget-ae9179fd64d656b378f2" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-ae9179fd64d656b378f2">{"x":{"filter":"top","vertical":false,"filterHTML":"<tr>\n  <td><\/td>\n  <td data-type=\"character\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"factor\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n    <div style=\"width: 100%; display: none;\">\n      <select multiple=\"multiple\" style=\"width: 100%;\" data-options=\"[&quot;1&quot;,&quot;10&quot;,&quot;11&quot;,&quot;12&quot;,&quot;13&quot;,&quot;14&quot;,&quot;15&quot;,&quot;16&quot;,&quot;17&quot;,&quot;18&quot;,&quot;19&quot;,&quot;2&quot;,&quot;3&quot;,&quot;4&quot;,&quot;5&quot;,&quot;6&quot;,&quot;7&quot;,&quot;8&quot;,&quot;9&quot;,&quot;X&quot;,&quot;Y&quot;]\"><\/select>\n    <\/div>\n  <\/td>\n  <td data-type=\"integer\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n    <div style=\"display: none;position: absolute;width: 200px;opacity: 1\">\n      <div data-min=\"701933\" data-max=\"149282940\"><\/div>\n      <span style=\"float: left;\"><\/span>\n      <span style=\"float: right;\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"number\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n    <div style=\"display: none;position: absolute;width: 200px;opacity: 1\">\n      <div data-min=\"0.382\" data-max=\"1\" data-scale=\"3\"><\/div>\n      <span style=\"float: left;\"><\/span>\n      <span style=\"float: right;\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"integer\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n    <div style=\"display: none;position: absolute;width: 200px;opacity: 1\">\n      <div data-min=\"701933\" data-max=\"149441020\"><\/div>\n      <span style=\"float: left;\"><\/span>\n      <span style=\"float: right;\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"number\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n    <div style=\"display: none;position: absolute;width: 200px;opacity: 1\">\n      <div data-min=\"0.15038\" data-max=\"71.98931\" data-scale=\"15\"><\/div>\n      <span style=\"float: left;\"><\/span>\n      <span style=\"float: right;\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"character\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"character\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"logical\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n    <div style=\"width: 100%; display: none;\">\n      <select multiple=\"multiple\" style=\"width: 100%;\" data-options=\"[&quot;true&quot;,&quot;false&quot;]\"><\/select>\n    <\/div>\n  <\/td>\n  <td data-type=\"logical\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n    <div style=\"width: 100%; display: none;\">\n      <select multiple=\"multiple\" style=\"width: 100%;\" data-options=\"[&quot;true&quot;,&quot;false&quot;]\"><\/select>\n    <\/div>\n  <\/td>\n  <td data-type=\"character\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"logical\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n    <div style=\"width: 100%; display: none;\">\n      <select multiple=\"multiple\" style=\"width: 100%;\" data-options=\"[&quot;true&quot;,&quot;false&quot;]\"><\/select>\n    <\/div>\n  <\/td>\n<\/tr>","data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46","47","48","49","50","51","52","53","54","55","56","57","58","59","60","61","62","63","64","65","66","67","68","69","70","71","72","73","74","75","76","77","78","79","80","81","82","83","84","85","86","87","88","89","90","91","92"],["backuprs30400977","backupUNC010151687","backupUNC010562539","backupUNC040068041","backupUNC040226607","backupUNC070042677","backupUNC070042912","backupUNC100008688","backupUNC100125042","backupUNC110001327","backupUNC120027003","backupUNC120289311","backupUNC130326039","backupUNC140022542","backupUNC140394528","backupUNC150434093","backupUNC160058135","backupUNC160212273","JAX00123350","JAX00633194","UNC_rs30519038","UNC_rs30619206","UNC_rs33690796","UNC_rs46376188","UNC_rs47034953","UNC_rs47365060","UNC_rs47503048","UNC_rs48613746","UNC_rs50749693","UNC_rs51459434","UNC_rs51822221","UNC010153451","UNC010484549","UNC020037344","UNC020056193","UNC020211155","UNC020307523","UNC020558243","UNC020771755","UNC021329832","UNC030161917","UNC030764068","UNC040129622","UNC040509337","UNC040626133","UNC050078382","UNC050377082","UNC050482226","UNC060165098","UNC060169921","UNC060244707","UNC060263578","UNC061457241","UNC061759927","UNC070330827","UNC070416673","UNC070623052","UNC070979841","UNC080157254","UNC080221309","UNC080452715","UNC090265546","UNC090303930","UNC090520271","UNC100158364","UNC100221671","UNC110010051","UNC110041950","UNC110160969","UNC120325121","UNC130066907","UNC130140599","UNC130330364","UNC130568787","UNC131295000","UNC140108037","UNC150193869","UNC150334867","UNC160149422","UNC160185764","UNC170010238","UNC170194414","UNC170211647","UNC170260163","UNC180228941","UNC181051157","UNC190204409","UNC200015082","UNC200036229","UNC200093471","UNC210000599","UNC210001275"],["8","1","1","4","4",null,"7","10","10","11","12","12","13","14","14","15","16","16","4","7","19","14","5","19","16","13","9","11","8","16","5","1","1",null,"2","2","2","2","2",null,"3","3","4","4",null,"5","5","5",null,"6","6","6","6","6","7","7","7","7",null,"8","8",null,"9",null,"10","10","11","11","11",null,"13","13","13",null,"13","14","15",null,"16","16",null,null,"17","17","18",null,null,"X","X",null,"Y","Y"],[16535086,78645591,51419402,54544170,112306624,null,21002474,3476917,107827071,5043505,17853404,35406554,86752474,12452676,51259721,56379245,44926011,41005689,112329365,20993306,55252994,22414612,96772409,50928720,80183085,4161655,115857055,62564867,23505750,37749179,40353626,79156169,50131703,null,51208277,144158217,44439999,4722557,149282940,null,103622867,9365596,74706152,54266944,null,29739235,63250213,98390366,null,108125744,136299318,138072154,114338312,135079513,77761632,116158933,18113419,31858520,null,52173058,82462989,null,104394791,null,49900407,89142252,9158114,37088832,99084561,null,54224280,79750665,87602678,null,111945838,28651885,53209725,null,94600830,78435904,null,null,83610236,84143405,80267053,null,null,61591827,129387280,null,1991009,701933],[0.764,0.436,0.636,0.4,0.673,1,1,0.491,0.382,1,0.418,0.455,1,1,1,1,1,0.455,0.673,0.436,0.618,0.564,0.4,0.6,0.618,1,0.436,0.436,0.418,0.455,0.618,1,0.6,0.509,0.473,1,0.564,0.527,0.582,0.618,0.455,0.436,1,0.4,1,0.455,0.4,0.491,0.655,0.636,0.4,0.491,0.655,0.4,0.564,0.491,1,0.4,0.509,0.527,0.6,0.709,0.564,0.382,1,1,0.473,0.418,1,0.818,0.745,1,0.6,0.691,0.473,0.564,1,1,0.418,0.564,0.527,1,0.436,0.818,0.545,0.782,0.455,0.8,0.4,1,0.382,0.545],[16485072,78667874,51380243,54544170,112449427,null,21268549,3526917,107991210,5093505,17803403,35356555,86604355,10017097,51022264,56515849,45105648,41185327,112472168,21259381,55264562,22364544,96624550,50940282,80386197,4111656,116027987,62674041,23015734,37928817,40196283,79178452,50092544,null,51318265,144316297,44549987,4717746,149441020,null,103715551,9300536,74787915,54266944,null,29534237,63092870,98242507,null,108148783,136322320,138095156,114361351,135102515,78111884,116559698,18379494,32159095,null,51720023,81736360,null,104517592,null,50024311,89306390,9208114,37198005,99193735,null,54070261,79602546,87454559,null,111809304,28929928,53346329,null,94799971,78639016,null,null,83302807,83835976,80223838,null,null,62548221,130486531,null,1991009,701933],[6.659087241199,39.3551733699617,25.012289690054,27.53059,50.26909,null,8.05787601087817,0.15038,54.4841869891859,1.22653011106441,6.57218993093798,14.14213,43.30448,1.71839260330291,20.1554192556322,21.3992956584151,27.4385606876801,26.18058,50.26909,8.05717776932836,48.4472838263573,6.09423,46.14742,42.22892,44.03492,0.653805177643575,66.2597998785594,36.7571812868168,9.61877,24.63626,19.54072,39.38927,24.35202,null,27.9952538352963,69.187899073276,25.4488545313999,1.22860544045982,71.98931,null,43.94374,0.738689139243125,35.07252,27.4335652005777,null,13.3832977476034,30.5707734745389,46.46069,null,48.0721726376877,64.8882926059681,66.6229110627125,51.39775,64.2345089292831,41.9972857879088,59.8010056696091,7.15542,17.3084407303981,null,27.365966854096,37.17231,null,54.56675,null,23.24554,43.2518128116971,4.12142,20.9224491037376,60.99679,null,26.64019,39.8451898573945,43.3877828806703,null,61.6735238380084,10.73888,18.9631829634759,null,53.6056180205002,43.50494,null,null,50.4547734095155,52.765691189997,51.3523402154757,null,null,33.01692,51.58803,null,null,null],["minus","minus","minus","minus","plus",null,"minus","minus","plus","minus","plus","minus","plus","plus","minus","plus","minus","minus","minus","minus","minus","minus","minus","minus","minus","minus","minus","minus","minus","minus","minus","minus","minus",null,"minus","minus","minus","minus","minus",null,"plus","minus","minus","plus",null,"minus","minus","plus",null,"minus","plus","minus","minus","plus","minus","minus","minus","plus",null,"minus","plus",null,"minus",null,"minus","plus","minus","plus","minus",null,"minus","plus","minus",null,"minus","minus","minus",null,"minus","minus",null,null,"plus","minus","minus",null,null,"minus","minus",null,"plus","minus"],["AG","AG","TC","TC","AG","AG","AG","AG","AG","TC","AG","TC","AG","TG","AG","AC","AG","TC","TC","AG","TC","TC","AG","TG","AC","AG","AG","AC","TC","AG","TC","AG","TC","AG","TC","TC","TC","TC","AC","TG","AG","TG","AG","AG","TG","AG","TG","AG","TC","TC","AG","TC","TC","AG","TC","TC","AG","AG","AC","AC","AG","TC","TG","TG","TG","AG","TC","AG","TG","TC","TC","AG","TC","TG","TC","TC","TG","TC","AC","TC","AG","AG","AG","TC","TC","AC","AC","TC","AG","AG","AG","TC"],[true,true,true,true,true,false,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,false,true,true,true,true,true,false,true,true,true,true,false,true,true,true,false,true,true,true,true,true,true,true,true,true,false,true,true,false,true,false,true,true,true,true,true,false,true,true,true,false,true,true,true,false,true,true,false,false,true,true,true,false,false,true,true,false,true,true],[false,false,false,false,false,true,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,true,false,false,false,false,false,true,false,false,false,false,true,false,false,false,false,false,false,false,false,false,false,false,false,false,true,false,false,false,false,true,false,false,false,false,false,true,false,false,false,true,false,false,false,true,false,false,false,true,false,false,false,true,true,false,false,true,false,false],["CTGTCAGATAAAAGTACTTAACAACCCTAAAGGGGCCAGAGCTCTTCTTT","ACTTATTTTTCAGTAGCCCAGATTCTGTCACTTTGTGCTTCCGTTATTCT","ACAACCATAATAACATTTATTACTAAGATTTTAGAAAATGTAAAAGTGTA","CACTTATTCTTTTTACATTTACTGGCTTTAATATGTTTTTTTTGGGGGGG","AAGTGGGTACTTAAAGGCCTATTGCTATCTGGTAATCCACTAAATATACA","TGTCCCCACCAACCAAGAATAGCCCCCCCCCATTGTAGTTTGGAAGATGA","CCCCTTACTGATTTTAATATGCTTTCTTTGTGTGTGTGTGTGTGTATTTG","CCAGGAAAGGAGTAAAGTACAGGCTTCAGGAGTAAAGTACAGGAAAGAGC","AAGAGATGCTGCAGTCAGAAACATAAGAACCACAAAAGGAAAAGAAAAAG","CAACCACTAGCTGCCTCTCTGCCTCCGGCGTGCTGGGATTAAAGGCATGC","TCCTCCCAGTTCCTCAGCTAGCTGTTAAAAGCCCCCTACTGTCACTGTTC","TGTGTTCATGCTGTGAACTGGAAAGCCTCTTTTTCATGATTCTGAATGTA","CAGCTTGGTCTACAACGTGACTTCTAAAACAACCTTTGAAAACCAAGGAT","GAGGAACACTGAATGGCTGAGAAGCACCTAAAAAATACTCAACATCCTTA","ACTTCTCTTTCCACGGGAGATAAACCACACCTTCAGCTTTTTGTTCTCTT","AGAAGTCAGTGCCTAGGGAAGCAACTAATTGCTTTGCCCATAGCAGGTAA","GTTATATTATTTTGTGTTAGTATTAATTAATAGGGGAAATGGCTTTCTAT","GATACTTATGCCATCAAATATACGTCACCTTTCAAAGTCACGTTGCCTTT","AAAGACTGAGAACAAATACTCATTTCATCATTCATATGTTCATTTCTCCA","CTGATGCAACCAATTTTCACTAAACAGTTTAATGTGTTAAATGGCCCTCA","CTGTAGCTCTTCCCAGGTCCAACAGGCCTGCACTCTGCTCAAAACCATCA","GTAAAAGCCACACCCACAAAACAGAAAGGCAGCATCTTGGCATTCAAGCA","TGGGTGCTGATGCAGACAGACAGTACACATGTGCAGATGCATGAACTATG","CAATAAAGCTTGTATATAATAATGAACAACAAAGAGAGGTAGTAAAAAAT","TCAGGATGATAGCATTGCCTTACAGATCTGAAATAAATCCAAGGGGTACT","CCAGGGCCAGATCCATCCATTTGGCCTAAGACAAAGTTTGGAGCTGTGAC","GGAGCACACTCAGTCTAGGTTCTCAACAAACCGTGACATTCTGAAGCCTC","AAAGGACTTGAATAAACACTTCTCCGAAGGAGAAATACAAGTGAACAATA","GTTACAAGGCGGATGCAACAGAGGACAAAACACTTCCTCAGTGTCTCTGG","GCTAAGTTGATGTCCTGAATTGGATTCCTGGGTTCCACAAGGTGATAACC","GGAAGTACAGCTTAGTTCATATTGTTGTTCCTCCTGTATGGATGCAGATA","AGCATCTTCTGAGTATATGCCCAGGAGTGGTATTGCTGGATCTTAGAGTC","TTTAGGCCAAAGAGTGCTAAACATAGTTCATATGTTTGATTTATTTTTAG","TCAAGAAACTTGAAAGGTCTGGGGGGGACAATAGGAGTGGATAGTTCAGT","CAATAGATGCCTACCTGGCTGTTGCCAGGTAACTGCACAAATGTCTCCCA","CAGGTCCTGTGGGTTCCATTTTCTTTCCTAGTTCTTGTTATGCTCGCCCG","CCTTCCTCCCAAGTTCTGGAATTAAAGACCTACATACACCACCACCGCCC","GGGAGTAGGGGCCGCGCCGTCGAAGGGGGCGGCGCCGCCGGGACCCCGGG","ACAGGATTATGGAGAACATTTCATTTTTAAATCATAAAATCATTGTTTTT","GAATAACAAAAAAAAAAAAAAAAAAAAACAAAAAAACAAAAAACAAAAAA","CTGTGCTTATGCATTCCCCTGGCCACTGTCACAGTCTAGTCTTTTCTGAA","AGTGCCCGCCCGCCCTGAGGACTTCAGCCTCCCCCACCCCAGCCCCCCCC","AGTATTTATTTCATTTACACTTCCAATGCTATCCCAAAAGCCCCCCACAT","TACCTGTATGGATACTGTGCTCTCAGAATTGATTGACAGCATCTTTAACA","ACGGTGAACATTTACCTCATGACCAAAAAAAAAAAAAAAAAAAAAAAAAA","ACTAAAATCTAAACATAATCATAATCCTTTAAATTAAAAATGATGATAGA","GTTGCCATTTTTTGTAGTATATTTTGAGGTCAACTAGTGTAATACTTCAG","GGAAAATAAGACTTAGAGAGCTTAGGACACTTGTTCCATGTTGGGGAGCC","TACAGAGTGAGTTCCAGGACAGCCAGGGCTACACAGAGAAACCCTGTCTC","GCAGTAGCTTCACTTCTAGGTCCGGGTAGCACTGAAACAGATAGCACTAC","GAAATTCCTACCGTCACATAATAATCAAAACATCAAATGTACAAAACAAA","TGCTGGCTTTTCCACTTGATAACTGTACCTGTGTGCTCTAAGGAGAAATG","CATGTGTGTTCCAAGCCTGATTTCACTGGTTTTTTTTTCCCCACCTTTTA","CTTTGTATTTATTGAACAGTTTAGAAAACAAGGTATAGGAGAGATATGAG","ATTACTAAGAAAACTCAACAACTTAGAATCTTATTTGTGTGTGTGTGTGT","AGGAAGCGGCAGAACTGCAGCTTCTCGGGTGAGGTGAGGAAGCTCTTGTA","AGTGAGGCTGGGTCATTCTGATATGGTGGCCCAAAGTACTTTAGAAATTC","ATTTAAGGCCATGTGTCTGTGATGAGGTATTGGAGGTTGCACCATTATGC","GGAAAAAAAATAAATGAATTGATTAAAAAAAGAAAATAGATTTTTCATAA","ACCAAAATAAACATAATGGGGGTGCAGACCTAAATAAGAATTCTCAAAAA","AAAGAGGTGCAGTCTCTAGACACCCACTCAGCTCTGTGGTGTTTTTTTTT","GTGAGGAGGGAATGCAAGAGTGAGTGAGGAGGGAATGTGAGAGTCAGTGA","ATAGAAGCAAAAGGCCCAAAGTTACGAGAGCTTTTTACGGTGGGATAGTA","TGTCTCAAAAAACCAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAC","ATGCCTTATACTTCAGTAAAAATTTCTGTATTGATGAATTCTAATTCCCT","ATGCTTTCTCAGCATCTAATGAAATTTTCATGTGGTTTTTTTTTCTTTGA","AAACAGGAAGCACAAACTCCCTAAAAGAATTATAGGAAAACACTGCTAAC","TTTATTGTTAACCCAAGTAGGGTTTCTCAGCTAAAGGATGTTCCTCTTGT","GAGTTATGGCAAAGGGACCTTTGAGTTAAAAAGCATAAGGTTTTTTTTGT","TGAAAATATCTCTTTCTCTCTCTCTCTCTCTCTCTCTCACACACACACAC","GTCCCTTTATACCTAGTACTGTCTTGCGTGAATTCTTTTACTGCCCAACC","TCTTCAGAGATGAGTTCTTGTCATACAGATCTTTCACTTCTTTAGTTAGA","ACAGTTGGCCAGAACTCCAGGGAGATGAGCTCTCTTCTCCTGGTACTTGA","ATATCTTAAACACAATATATGTAATATATGAAAAATATATAAATAATTGA","GACAGACCGTGAGTCCTCCGGCTTGCGTCCCCTTCCAGAACCAGGGACTC","TCTCCTTACTTCCCTACTTACACTCAAATGTATAACCTTCTTCTTTAGTT","GGTGGGCGACGCCCGGCCGGCTAGCGCTCCGCCCGCCGCCCGTGCGCTCC","AAGAAAGAAAGAAAGAAAGAAAGAAAGAAAGAAAGAAAGAAAGAAAGAAA","TGCACACATGACATCATACTCATGACATTGTACACACATGTGACATCATA","TAAAGATTGTGCCTCATCTTTTATTTTAGTTTCTAATTTCCAATGCATTA","TTAGGACCATTAATCACATTTGTCTTCTGACCAGGCGGCATGGCTTCCTT","GTGTGTATATATTCATATGTGTATCATATGTATATGATATATACATATAC","AATTAATTTTATTGCATTAGTGTTAAGTCAGGAATGCACATGCCTGAGAG","GGGTGGGGGTGGGGATAAGGGATTGGGGGTGTGTGCTGCACTGGCTGCTG","CTTCTGCCTCCCACGTGCTGGTGTTAATTAAAGGCACGCACCACCATGTT","GTTAAATACCTAATAAAAAATGGAAAAAAAAAAAAAGAAATACACCCCCC","CTCCCCCAAACCCCTTCTTGTTTTTTTGTTTTTTGTTTGTTTCTTTTTTT","TGGTGTCCTGCTGGAGTTCTTGTGTCTGCAGAGCTTTTCATTCAGCCCCA","AAGCATTGTAGGCCATCATATAGTATGTGGTTGTTTTCTTAAAATTCATG","AATATATAGAACTCATATATACATATATGTATATATTTAAGAGAGTTATA","CAGAACAGAGGTGTCCACCCTGCCTGGGAGGGCATTTTTGGAGCATCTGA","CGACCCCGACCCAAACTTTTTACAGGCTTATAAAGGCAAAAACCACAAAA"],[false,false,true,true,false,false,false,false,false,true,false,true,false,false,false,false,false,false,false,true,false,false,false,false,false,false,false,false,false,false,false,false,false,false,true,false,true,true,false,false,false,true,false,false,false,false,false,false,false,true,false,false,false,false,true,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,true,false,true,false,false,false,false,true,false,false,false,true,false,false,false,true,false,false,false,true]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>marker<\/th>\n      <th>chr<\/th>\n      <th>bp_grcm39<\/th>\n      <th>freq<\/th>\n      <th>bp_mm10<\/th>\n      <th>cM_cox<\/th>\n      <th>strand<\/th>\n      <th>snp<\/th>\n      <th>unique<\/th>\n      <th>unmapped<\/th>\n      <th>probe<\/th>\n      <th>strand_flipped<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"width":"20%","targets":8},{"className":"dt-right","targets":[3,4,5,6]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false,"orderCellsTop":true}},"evals":[],"jsHooks":[]}</script>
<p><img src="figure/MUGA_Reference_QC.Rmd/Plotting%20no%20calls-2.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-no-spaces-Plotting-no-calls-2">
Past versions of "Plotting no calls-2.png"
</button>
</p>
<div id="fig-no-spaces-Plotting-no-calls-2" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/sam-widmayer/MUGA_reference_data/blob/3501596652f350afae3250facfce0abffc9ee2f5/docs/figure/MUGA_Reference_QC.Rmd/Plotting no calls-2.png" target="_blank">3501596</a>
</td>
<td>
sam-widmayer
</td>
<td>
2022-11-09
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="sample-qc" class="section level2">
<h2>Sample QC</h2>
<div id="searching-for-samples-with-poor-marker-representation"
class="section level3">
<h3>Searching for samples with poor marker representation</h3>
<p>In a similar fashion, we calculated the number of reference samples
with missing genotypes. Repeated observations of samples/strains with
identical names meant that genotype counts for each marker among them
couldn’t be grouped and tallied, so determining no-call frequency
occurred column-wise. Mouse over individual samples to see the number of
markers with missing genotypes for each sample.</p>
<pre class="r"><code>## Calculating the number of missing markers for each sample
n.calls.strains.df &lt;- tidy_control_genotypes %&gt;%
  dplyr::select(sample, marker, genotype) %&gt;%
  dplyr::group_by(sample, genotype) %&gt;%
  dplyr::count() %&gt;%
  dplyr::filter(genotype == &quot;N&quot;) %&gt;%
  dplyr::ungroup() %&gt;%
  dplyr::select(sample, n) %&gt;%
  dplyr::rename(n.no.calls = n)
#    sample              n.no.calls
#    &lt;chr&gt;                    &lt;int&gt;
#  1 #1-m-9376 (m)              149
#  2 #2-m-9376 (m)              201
#  3 #3-m-9376 (m)              191
#  4 #4-m-9376 (m)              226
#  5 #5-m-9376 (m)              217
#  6 #6-m-9376 (m)              154
#  7 129S1/SvImJ (f)            176
#  8 129S1/SvImJm212 (m)        158
#  9 129xCASTf005 (f)           113
# 10 129xPWK 040F (f)           108


## Interactive plot of the number of missing genotypes for each sample.
sampleQC &lt;- ggplot(n.calls.strains.df, 
                   mapping = aes(x = reorder(sample,n.no.calls), 
                                 y = n.no.calls,
                                 text = paste(&quot;Sample:&quot;, sample))) + 
  geom_point() +
  QCtheme + 
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank()) +
  labs(x = &quot;Number of mice with missing genotypes&quot;,
       y = &quot;Number of markers&quot;)
ggplotly(sampleQC, tooltip = c(&quot;text&quot;,&quot;y&quot;))</code></pre>
<div id="htmlwidget-075775afc8c85fc742ec" style="width:672px;height:480px;" class="plotly html-widget"></div>
<script type="application/json" data-for="htmlwidget-075775afc8c85fc742ec">{"x":{"data":[{"x":[23,32,30,35,34,24,27,25,11,9,4,22,21,40,33,39,6,7,3,2,8,49,36,37,1,42,50,44,12,5,10,13,26,18,20,46,41,53,48,38,55,52,47,43,54,16,14,29,28,51,45,31,17,15,19],"y":[149,201,191,226,217,154,176,158,113,108,102,144,142,292,202,276,106,106,73,65,107,326,250,270,51,302,326,310,116,105,109,117,175,125,133,316,297,372,323,272,394,367,319,307,391,122,119,179,177,361,312,191,122,121,125],"text":["n.no.calls: 149<br />Sample: #1-m-9376 (m)","n.no.calls: 201<br />Sample: #2-m-9376 (m)","n.no.calls: 191<br />Sample: #3-m-9376 (m)","n.no.calls: 226<br />Sample: #4-m-9376 (m)","n.no.calls: 217<br />Sample: #5-m-9376 (m)","n.no.calls: 154<br />Sample: #6-m-9376 (m)","n.no.calls: 176<br />Sample: 129S1/SvImJ (f)","n.no.calls: 158<br />Sample: 129S1/SvImJm212 (m)","n.no.calls: 113<br />Sample: 129xCASTf005 (f)","n.no.calls: 108<br />Sample: 129xPWK 040F (f)","n.no.calls: 102<br />Sample: 129xPWK037m (m)","n.no.calls: 144<br />Sample: A/J (f)","n.no.calls: 142<br />Sample: A/Jm111 (m)","n.no.calls: 292<br />Sample: AB0095 (f)","n.no.calls: 202<br />Sample: AE-F (f)","n.no.calls: 276<br />Sample: AH0030 (f)","n.no.calls: 106<br />Sample: AJxPWKf001 (f)","n.no.calls: 106<br />Sample: AJxPWKm006 (m)","n.no.calls:  73<br />Sample: B6xPWKf002 (f)","n.no.calls:  65<br />Sample: B6xPWKm006 (m)","n.no.calls: 107<br />Sample: B6xWSB 15713 (m)","n.no.calls: 326<br />Sample: BD0018 (f)","n.no.calls: 250<br />Sample: BE0016 (f)","n.no.calls: 270<br />Sample: BF0016 (f)","n.no.calls:  51<br />Sample: C57BL/6J (f)","n.no.calls: 302<br />Sample: CA0056 (f)","n.no.calls: 326<br />Sample: CAST/EiJ (f)","n.no.calls: 310<br />Sample: CAST/EiJm42 (m)","n.no.calls: 116<br />Sample: CASTx129m001 (m)","n.no.calls: 105<br />Sample: CASTxAJf001 (f)","n.no.calls: 109<br />Sample: CASTxAJm005 (m)","n.no.calls: 117<br />Sample: CASTxNODf007 (f)","n.no.calls: 175<br />Sample: CASTxPWKm021 (m)","n.no.calls: 125<br />Sample: CASTxWSBf015 (f)","n.no.calls: 133<br />Sample: CASTxWSBm001 (m)","n.no.calls: 316<br />Sample: CB15916F (f)","n.no.calls: 297<br />Sample: CD0063 (f)","n.no.calls: 372<br />Sample: CE0005 (f)","n.no.calls: 323<br />Sample: CH0013 (f)","n.no.calls: 272<br />Sample: DA0018 (f)","n.no.calls: 394<br />Sample: DE0001 (f)","n.no.calls: 367<br />Sample: DH0141 (f)","n.no.calls: 319<br />Sample: EH15876F (f)","n.no.calls: 307<br />Sample: NOD/LtJ (f)","n.no.calls: 391<br />Sample: NOD/ShiLtJ-35335 (m)","n.no.calls: 122<br />Sample: NODxPWKm003 (m)","n.no.calls: 119<br />Sample: NODxPWKm004 (f)","n.no.calls: 179<br />Sample: NZO/HlLtJ (f)","n.no.calls: 177<br />Sample: NZO/HlLtJm51 (m)","n.no.calls: 361<br />Sample: PWK/PhJ (f)","n.no.calls: 312<br />Sample: PWK/PhJm175 (m)","n.no.calls: 191<br />Sample: WSB/EiJ (f)","n.no.calls: 122<br />Sample: WSBxCASTf003 (f)","n.no.calls: 121<br />Sample: WSBxPWKf003 (f)","n.no.calls: 125<br />Sample: WSBxPWKm001 (m)"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(0,0,0,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null}],"layout":{"margin":{"t":26.2283105022831,"r":7.30593607305936,"b":28.4931506849315,"l":43.1050228310502},"plot_bgcolor":"rgba(255,255,255,1)","paper_bgcolor":"rgba(255,255,255,1)","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187},"xaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[0.4,55.6],"tickmode":"array","ticktext":["C57BL/6J (f)","B6xPWKm006 (m)","B6xPWKf002 (f)","129xPWK037m (m)","CASTxAJf001 (f)","AJxPWKf001 (f)","AJxPWKm006 (m)","B6xWSB 15713 (m)","129xPWK 040F (f)","CASTxAJm005 (m)","129xCASTf005 (f)","CASTx129m001 (m)","CASTxNODf007 (f)","NODxPWKm004 (f)","WSBxPWKf003 (f)","NODxPWKm003 (m)","WSBxCASTf003 (f)","CASTxWSBf015 (f)","WSBxPWKm001 (m)","CASTxWSBm001 (m)","A/Jm111 (m)","A/J (f)","#1-m-9376 (m)","#6-m-9376 (m)","129S1/SvImJm212 (m)","CASTxPWKm021 (m)","129S1/SvImJ (f)","NZO/HlLtJm51 (m)","NZO/HlLtJ (f)","#3-m-9376 (m)","WSB/EiJ (f)","#2-m-9376 (m)","AE-F (f)","#5-m-9376 (m)","#4-m-9376 (m)","BE0016 (f)","BF0016 (f)","DA0018 (f)","AH0030 (f)","AB0095 (f)","CD0063 (f)","CA0056 (f)","NOD/LtJ (f)","CAST/EiJm42 (m)","PWK/PhJm175 (m)","CB15916F (f)","EH15876F (f)","CH0013 (f)","BD0018 (f)","CAST/EiJ (f)","PWK/PhJ (f)","DH0141 (f)","CE0005 (f)","NOD/ShiLtJ-35335 (m)","DE0001 (f)"],"tickvals":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55],"categoryorder":"array","categoryarray":["C57BL/6J (f)","B6xPWKm006 (m)","B6xPWKf002 (f)","129xPWK037m (m)","CASTxAJf001 (f)","AJxPWKf001 (f)","AJxPWKm006 (m)","B6xWSB 15713 (m)","129xPWK 040F (f)","CASTxAJm005 (m)","129xCASTf005 (f)","CASTx129m001 (m)","CASTxNODf007 (f)","NODxPWKm004 (f)","WSBxPWKf003 (f)","NODxPWKm003 (m)","WSBxCASTf003 (f)","CASTxWSBf015 (f)","WSBxPWKm001 (m)","CASTxWSBm001 (m)","A/Jm111 (m)","A/J (f)","#1-m-9376 (m)","#6-m-9376 (m)","129S1/SvImJm212 (m)","CASTxPWKm021 (m)","129S1/SvImJ (f)","NZO/HlLtJm51 (m)","NZO/HlLtJ (f)","#3-m-9376 (m)","WSB/EiJ (f)","#2-m-9376 (m)","AE-F (f)","#5-m-9376 (m)","#4-m-9376 (m)","BE0016 (f)","BF0016 (f)","DA0018 (f)","AH0030 (f)","AB0095 (f)","CD0063 (f)","CA0056 (f)","NOD/LtJ (f)","CAST/EiJm42 (m)","PWK/PhJm175 (m)","CB15916F (f)","EH15876F (f)","CH0013 (f)","BD0018 (f)","CAST/EiJ (f)","PWK/PhJ (f)","DH0141 (f)","CE0005 (f)","NOD/ShiLtJ-35335 (m)","DE0001 (f)"],"nticks":null,"ticks":"","tickcolor":null,"ticklen":3.65296803652968,"tickwidth":0,"showticklabels":false,"tickfont":{"color":null,"family":null,"size":0},"tickangle":-0,"showline":false,"linecolor":null,"linewidth":0,"showgrid":false,"gridcolor":null,"gridwidth":0,"zeroline":false,"anchor":"y","title":{"text":"Number of mice with missing genotypes","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187}},"hoverformat":".2f"},"yaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[33.85,411.15],"tickmode":"array","ticktext":["100","200","300","400"],"tickvals":[100,200,300,400],"categoryorder":"array","categoryarray":["100","200","300","400"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.65296803652968,"tickwidth":0.66417600664176,"showticklabels":true,"tickfont":{"color":"rgba(0,0,0,1)","family":"","size":11.689497716895},"tickangle":-0,"showline":false,"linecolor":null,"linewidth":0,"showgrid":false,"gridcolor":null,"gridwidth":0,"zeroline":false,"anchor":"x","title":{"text":"Number of markers","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187}},"hoverformat":".2f"},"shapes":[{"type":"rect","fillcolor":"transparent","line":{"color":"rgba(51,51,51,1)","width":0.66417600664176,"linetype":"solid"},"yref":"paper","xref":"paper","x0":0,"x1":1,"y0":0,"y1":1}],"showlegend":false,"legend":{"bgcolor":"rgba(255,255,255,1)","bordercolor":"transparent","borderwidth":1.88976377952756,"font":{"color":"rgba(0,0,0,1)","family":"","size":11.689497716895}},"hovermode":"closest","barmode":"relative"},"config":{"doubleClick":"reset","modeBarButtonsToAdd":["hoverclosest","hovercompare"],"showSendToCloud":false},"source":"A","attrs":{"80cc5449185a":{"x":{},"y":{},"text":{},"type":"scatter"}},"cur_data":"80cc5449185a","visdat":{"80cc5449185a":["function (y) ","x"]},"highlight":{"on":"plotly_click","persistent":false,"dynamic":false,"selectize":false,"opacityDim":0.2,"selected":{"opacity":1},"debounce":0},"shinyEvents":["plotly_hover","plotly_click","plotly_selected","plotly_relayout","plotly_brushed","plotly_brushing","plotly_clickannotation","plotly_doubleclick","plotly_deselect","plotly_afterplot","plotly_sunburstclick"],"base_url":"https://plot.ly"},"evals":[],"jsHooks":[]}</script>
<p>From the sample QC plot, we can see that one of the NOD/ShiLtJ
references samples is missing 391 genotypes, reducing our confidence in
its use for consensus building. We exclude this sample in subsequent
analyses.</p>
</div>
<div id="validating-sex-of-reference-samples" class="section level3">
<h3>Validating sex of reference samples</h3>
<p>We next validated the sexes of each sample using sex chromosome probe
intensities. We paired up probe intensities, joined available metadata,
and filtered down to only markers covering the X and Y chromosomes.</p>
<pre class="r"><code>## Bad samples
bad_samples &lt;- c(&quot;NOD/ShiLtJ-35335 (m)&quot;)

## Reading in genotype intensities
long_intensities &lt;- tidy_control_genotypes %&gt;%
  dplyr::select(sample, marker, Chromosome, Position, genotype, X, Y, theta)

## Joining slimmer intensity files with marker metadata and reducing to markers on sex chromosomes
long_XY_intensities &lt;- long_intensities %&gt;%
  dplyr::left_join(., muga_metadata) %&gt;%
  dplyr::filter(Chromosome %in% c(&quot;X&quot;,&quot;Y&quot;),
                !sample %in% bad_samples)

# Expected output
#          marker Chromosome Position genotype     X     Y  chr bp_mm10 bp_grcm39   cM_cox strand snp unique unmapped
# 1: UNC200279739          X  4324915        H 0.745 1.560 &lt;NA&gt;      NA        NA       NA   &lt;NA&gt;  AG  FALSE    FALSE
# 2: UNC200000454          X  5080366        T 0.903 0.019    X 5503248   5415302 1.131211  minus  AC   TRUE    FALSE
# 3: UNC200001507          X  5282192        H 1.791 1.417 &lt;NA&gt;      NA        NA       NA   &lt;NA&gt;  AG  FALSE    FALSE
# 4: UNC200282646          X  5667495        A 0.446 0.010    X 6090377   6002431 1.406192   plus  AG   TRUE    FALSE
# 5: UNC200880603          X  5938298        G 0.016 0.517    X 6361180   6273234 1.533023  minus  TC   TRUE    FALSE
# 6: UNC200285015          X  6105555        A 0.998 0.026    X 6528437   6440491 1.611357   plus  AG   TRUE    FALSE
#                                                 probe strand_flipped
# 1: ATTCTTTGAATTTGTTCAGTGTCTCTTGTTAGGACTGTACTTTCATTTCT          FALSE
# 2: AGTGATACATACTCACATTCAGCTTCCAGTGTGATACTGCAGGGTGAAAC          FALSE
# 3: TATTTTAGTCTATTGTATAGTCCACCTTTCCCCTAGGCCATTGTAAATAC          FALSE
# 4: AGCAAGACATAATTTCCACATTACTAAGTCAATTAACGTGCAAATATGAC          FALSE
# 5: ATCTTGTTCAGATGTGTTATCTGGATAGTGCTGGCTTCACAGACTAAGTT          FALSE
# 6: GCAGCCTGCAAAACACCTTCCAAAACTACAAAAGCTGGCTAACAGAGAGC          FALSE</code></pre>
<p>Then we flagged markers with high missingness across all samples, as
well as samples with high missingness among all markers.</p>
<pre class="r"><code>## Flagging markers and samples based on previous QC steps
flagged_XY_intensities &lt;- long_XY_intensities %&gt;%
  dplyr::mutate(marker_flag = dplyr::if_else(condition = marker %in% above.cutoff$marker,
                                             true = &quot;FLAG&quot;,
                                             false = &quot;&quot;))</code></pre>
<p>The first round of inferring predicted sexes used a rough search of
the sample name for expected nomenclature convention, which includes a
sex denotation.</p>
<pre class="r"><code># Vector of founder strain names
loose_founder_strains &lt;- c(&quot;A/J&quot;,&quot;C57BL/6J&quot;,&quot;129&quot;,&quot;NOD&quot;,
                           &quot;NZO&quot;,&quot;CAST&quot;,    &quot;PWK&quot;,&quot;WSB&quot;)

# Re-flag genotypes based on bad markers or bad samples.
# Inputs:
# 1) All sample genotypes
# 2) marker metadata
# 3) flag cutoff tables
genos.flagged &lt;- tidy_control_genotypes %&gt;%
  dplyr::left_join(., muga_metadata) %&gt;%
  # Flagging markers and samples
  dplyr::mutate(marker_flag = dplyr::if_else(condition = marker %in% above.cutoff$marker,
                                             true = &quot;FLAG&quot;,
                                             false = &quot;&quot;))

## First round of predicted sex inference
## Input: flagged XY intensities
prelim.predicted.sexes &lt;- flagged_XY_intensities %&gt;%
  dplyr::mutate(predicted.sex = dplyr::case_when(stringr::str_detect(string = sample, pattern = &quot;(f)&quot;) == TRUE ~ &quot;f&quot;,
                                                 stringr::str_detect(string = sample, pattern = &quot;(m)&quot;) == TRUE ~ &quot;m&quot;,
                                      TRUE ~ &quot;unknown&quot;))
## Output: flagged intensities with preliminary sex predictions and background assignments among F1 hybrids
#           sample       marker Chromosome Position genotype     X     Y  chr bp_mm10 bp_grcm39   cM_cox strand snp unique unmapped
# 1: #1-m-9376 (m) UNC200279739          X  4324915        H 0.745 1.560 &lt;NA&gt;      NA        NA       NA   &lt;NA&gt;  AG  FALSE    FALSE
# 2: #1-m-9376 (m) UNC200000454          X  5080366        T 0.903 0.019    X 5503248   5415302 1.131211  minus  AC   TRUE    FALSE
# 3: #1-m-9376 (m) UNC200001507          X  5282192        H 1.791 1.417 &lt;NA&gt;      NA        NA       NA   &lt;NA&gt;  AG  FALSE    FALSE
# 4: #1-m-9376 (m) UNC200282646          X  5667495        A 0.446 0.010    X 6090377   6002431 1.406192   plus  AG   TRUE    FALSE
# 5: #1-m-9376 (m) UNC200880603          X  5938298        G 0.016 0.517    X 6361180   6273234 1.533023  minus  TC   TRUE    FALSE
# 6: #1-m-9376 (m) UNC200285015          X  6105555        A 0.998 0.026    X 6528437   6440491 1.611357   plus  AG   TRUE    FALSE
#                                                 probe strand_flipped marker_flag predicted.sex
# 1: ATTCTTTGAATTTGTTCAGTGTCTCTTGTTAGGACTGTACTTTCATTTCT          FALSE                         m
# 2: AGTGATACATACTCACATTCAGCTTCCAGTGTGATACTGCAGGGTGAAAC          FALSE                         m
# 3: TATTTTAGTCTATTGTATAGTCCACCTTTCCCCTAGGCCATTGTAAATAC          FALSE                         m
# 4: AGCAAGACATAATTTCCACATTACTAAGTCAATTAACGTGCAAATATGAC          FALSE                         m
# 5: ATCTTGTTCAGATGTGTTATCTGGATAGTGCTGGCTTCACAGACTAAGTT          FALSE                         m
# 6: GCAGCCTGCAAAACACCTTCCAAAACTACAAAAGCTGGCTAACAGAGAGC          FALSE                         m



# Join the sample table with resex information with each sample&#39;s strain background from initial sex prediction.
# Inputs:
# 1) Sample metadata, including sex
# 2) Sample strain background
sample.meta &lt;- prelim.predicted.sexes %&gt;%
  dplyr::ungroup() %&gt;%
  dplyr::distinct(sample, predicted.sex)
#           sample predicted.sex
# 1: #1-m-9376 (m)             m
# 2: #3-m-9376 (m)             m
# 3: #5-m-9376 (m)             m
# 4: #2-m-9376 (m)             m
# 5: #4-m-9376 (m)             m
# 6: #6-m-9376 (m)             m

# From the sample metadata, extract any sample derived from an CC/DO founder.
founder_samples &lt;- sample.meta %&gt;%
  dplyr::mutate(founder = case_when(str_detect(sample, loose_founder_strains[1]) == TRUE ~ &quot;FOUNDER&quot;,
                                    str_detect(sample, loose_founder_strains[2]) == TRUE ~ &quot;FOUNDER&quot;,
                                    str_detect(sample, loose_founder_strains[3]) == TRUE ~ &quot;FOUNDER&quot;,
                                    str_detect(sample, loose_founder_strains[4]) == TRUE ~ &quot;FOUNDER&quot;,
                                    str_detect(sample, loose_founder_strains[5]) == TRUE ~ &quot;FOUNDER&quot;,
                                    str_detect(sample, loose_founder_strains[6]) == TRUE ~ &quot;FOUNDER&quot;,
                                    str_detect(sample, loose_founder_strains[7]) == TRUE ~ &quot;FOUNDER&quot;,
                                    str_detect(sample, loose_founder_strains[8]) == TRUE ~ &quot;FOUNDER&quot;,
                                    TRUE ~ &quot;NOT CC/DO Founder&quot;)) %&gt;%
  dplyr::filter(founder == &quot;FOUNDER&quot;)
#                 sample predicted.sex founder
# 1:             A/J (f)             f FOUNDER
# 2:         A/Jm111 (m)             m FOUNDER
# 3:        C57BL/6J (f)             f FOUNDER
# 4:     129S1/SvImJ (f)             f FOUNDER
# 5: 129S1/SvImJm212 (m)             m FOUNDER
# 6:         NOD/LtJ (f)             f FOUNDER


remaining_samples &lt;- sample.meta %&gt;%
  dplyr::filter(!sample %in% founder_samples$sample)
founder_letters &lt;- LETTERS[1:8]
recoded_founder_samples &lt;- remaining_samples %&gt;%
  dplyr::mutate(code = str_sub(start = 1, end = 2, string = sample),
                dam = str_sub(start = 1, end = 1, string = code),
                sire = str_sub(start = 2, end = 2, string = code)) %&gt;%
  dplyr::mutate(dam = case_when(dam == &quot;A&quot; ~ loose_founder_strains[1],
                                dam == &quot;B&quot; ~ loose_founder_strains[2],
                                dam == &quot;C&quot; ~ loose_founder_strains[3],
                                dam == &quot;D&quot; ~ loose_founder_strains[4],
                                dam == &quot;E&quot; ~ loose_founder_strains[5],
                                dam == &quot;F&quot; ~ loose_founder_strains[6],
                                dam == &quot;G&quot; ~ loose_founder_strains[7],
                                dam == &quot;H&quot; ~ loose_founder_strains[8]),
                sire = case_when(sire == &quot;A&quot; ~ loose_founder_strains[1],
                                 sire == &quot;B&quot; ~ loose_founder_strains[2],
                                 sire == &quot;C&quot; ~ loose_founder_strains[3],
                                 sire == &quot;D&quot; ~ loose_founder_strains[4],
                                 sire == &quot;E&quot; ~ loose_founder_strains[5],
                                 sire == &quot;F&quot; ~ loose_founder_strains[6],
                                 sire == &quot;G&quot; ~ loose_founder_strains[7],
                                 sire == &quot;H&quot; ~ loose_founder_strains[8]),
                founder = case_when(str_detect(dam, loose_founder_strains[1]) == TRUE ~ &quot;FOUNDER&quot;,
                                    str_detect(dam, loose_founder_strains[2]) == TRUE ~ &quot;FOUNDER&quot;,
                                    str_detect(dam, loose_founder_strains[3]) == TRUE ~ &quot;FOUNDER&quot;,
                                    str_detect(dam, loose_founder_strains[4]) == TRUE ~ &quot;FOUNDER&quot;,
                                    str_detect(dam, loose_founder_strains[5]) == TRUE ~ &quot;FOUNDER&quot;,
                                    str_detect(dam, loose_founder_strains[6]) == TRUE ~ &quot;FOUNDER&quot;,
                                    str_detect(dam, loose_founder_strains[7]) == TRUE ~ &quot;FOUNDER&quot;,
                                    str_detect(dam, loose_founder_strains[8]) == TRUE ~ &quot;FOUNDER&quot;,
                                    TRUE ~ &quot;NOT CC/DO Founder&quot;))



# Identify parental strains of founder samples
final_founder_samples &lt;- founder_samples %&gt;%
  tidyr::separate(sample, sep = &quot;x&quot;, into = c(&quot;dam&quot;,&quot;sire&quot;), remove = F) %&gt;% 
  dplyr::mutate(dam = case_when(str_detect(dam, loose_founder_strains[1]) == TRUE ~ loose_founder_strains[1],
                                str_detect(dam, loose_founder_strains[2]) == TRUE ~ loose_founder_strains[2],
                                str_detect(dam, loose_founder_strains[3]) == TRUE ~ loose_founder_strains[3],
                                str_detect(dam, loose_founder_strains[4]) == TRUE ~ loose_founder_strains[4],
                                str_detect(dam, loose_founder_strains[5]) == TRUE ~ loose_founder_strains[5],
                                str_detect(dam, loose_founder_strains[6]) == TRUE ~ loose_founder_strains[6],
                                str_detect(dam, loose_founder_strains[7]) == TRUE ~ loose_founder_strains[7],
                                str_detect(dam, loose_founder_strains[8]) == TRUE ~ loose_founder_strains[8],
                                str_detect(dam, &quot;AJ&quot;) == TRUE ~ loose_founder_strains[1],
                                str_detect(dam, &quot;B6&quot;) == TRUE ~ loose_founder_strains[2],
                                TRUE ~ &quot;NOT CC/DO Founder&quot;),
                sire = case_when(str_detect(sire, loose_founder_strains[1]) == TRUE ~ loose_founder_strains[1],
                                str_detect(sire, loose_founder_strains[2]) == TRUE ~ loose_founder_strains[2],
                                str_detect(sire, loose_founder_strains[3]) == TRUE ~ loose_founder_strains[3],
                                str_detect(sire, loose_founder_strains[4]) == TRUE ~ loose_founder_strains[4],
                                str_detect(sire, loose_founder_strains[5]) == TRUE ~ loose_founder_strains[5],
                                str_detect(sire, loose_founder_strains[6]) == TRUE ~ loose_founder_strains[6],
                                str_detect(sire, loose_founder_strains[7]) == TRUE ~ loose_founder_strains[7],
                                str_detect(sire, loose_founder_strains[8]) == TRUE ~ loose_founder_strains[8],
                                str_detect(sire, &quot;AJ&quot;) == TRUE ~ loose_founder_strains[1],
                                str_detect(sire, &quot;B6&quot;) == TRUE ~ loose_founder_strains[2],
                                TRUE ~ &quot;NOT CC/DO Founder&quot;)) %&gt;%
  dplyr::bind_rows(., recoded_founder_samples) %&gt;%
  dplyr::mutate(dam = as.factor(dam),
                sire = as.factor(sire)) %&gt;%
  dplyr::filter(!is.na(dam)) %&gt;%
  dplyr::select(-code) %&gt;%
  dplyr::mutate(sire = if_else(sire == &quot;NOT CC/DO Founder&quot;, dam, sire),
                bg = if_else(dam == sire, &quot;INBRED&quot;, &quot;CROSS&quot;))</code></pre>
<pre><code>Warning: Expected 2 pieces. Missing pieces filled with `NA` in 13 rows [1, 2, 3,
4, 5, 6, 7, 8, 9, 10, 11, 12, 13].</code></pre>
<pre class="r"><code>strict_founder_strains &lt;- c(&quot;129S1/SvImJ&quot;,&quot;A/J&quot;,&quot;C57BL/6J&quot;,&quot;CAST/EiJ&quot;,&quot;NOD/ShiLtJ&quot;,&quot;NZO/HlLtJ&quot;,&quot;PWK/PhJ&quot;,&quot;WSB/EiJ&quot;)
levels(final_founder_samples$dam) &lt;- strict_founder_strains
levels(final_founder_samples$sire)&lt;- strict_founder_strains

#                 sample         dam        sire predicted.sex founder     bg
# 1:             A/J (f)         A/J         A/J             f FOUNDER INBRED
# 2:         A/Jm111 (m)         A/J         A/J             m FOUNDER INBRED
# 3:        C57BL/6J (f)    C57BL/6J    C57BL/6J             f FOUNDER INBRED
# 4:     129S1/SvImJ (f) 129S1/SvImJ 129S1/SvImJ             f FOUNDER INBRED
# 5: 129S1/SvImJm212 (m) 129S1/SvImJ 129S1/SvImJ             m FOUNDER INBRED
# 6:         NOD/LtJ (f)  NOD/ShiLtJ  NOD/ShiLtJ             f FOUNDER INBRED


# Count up samples for each founder and resulting cross and display a table
# Dam names = row names; Sire name = column names
founder_sample_table &lt;- final_founder_samples %&gt;%
  dplyr::group_by(dam,sire) %&gt;%
  dplyr::count() %&gt;%
  tidyr::pivot_wider(names_from = sire, values_from = n)
DT::datatable(founder_sample_table, escape = FALSE, 
              options = list(columnDefs = list(list(width = &#39;20%&#39;, targets = c(8)))))</code></pre>
<div id="htmlwidget-bfd62b2a69b0da6666bc" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-bfd62b2a69b0da6666bc">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8"],["129S1/SvImJ","A/J","C57BL/6J","CAST/EiJ","NOD/ShiLtJ","NZO/HlLtJ","PWK/PhJ","WSB/EiJ"],[2,null,null,1,null,null,null,null],[1,2,null,2,1,null,null,null],[1,1,1,null,null,null,null,null],[1,null,1,2,null,null,null,1],[1,null,1,1,1,null,null,null],[1,1,1,null,1,2,null,null],[2,2,2,1,2,null,2,2],[1,1,1,2,1,1,null,1]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>dam<\/th>\n      <th>129S1/SvImJ<\/th>\n      <th>A/J<\/th>\n      <th>C57BL/6J<\/th>\n      <th>CAST/EiJ<\/th>\n      <th>NOD/ShiLtJ<\/th>\n      <th>NZO/HlLtJ<\/th>\n      <th>PWK/PhJ<\/th>\n      <th>WSB/EiJ<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"width":"20%","targets":8},{"className":"dt-right","targets":[2,3,4,5,6,7,8,9]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
<pre class="r"><code>XY_hetrate &lt;- long_XY_intensities %&gt;%
  # dplyr::right_join(., prelim.predicted.sexes, by = &quot;sample&quot;) %&gt;%
  dplyr::mutate(het = if_else(genotype == &quot;H&quot;, &quot;HET&quot;,&quot;NO_HET&quot;)) %&gt;%
  dplyr::select(sample, marker, Chromosome, genotype, het) %&gt;%
  dplyr::right_join(., prelim.predicted.sexes %&gt;%
                      dplyr::select(sample, predicted.sex)) %&gt;%
  dplyr::group_by(predicted.sex, Chromosome, het, sample) %&gt;%
  dplyr::count() %&gt;%
  tidyr::pivot_wider(names_from = het, values_from = n) %&gt;%
  dplyr::mutate(het_rate = HET/(HET+NO_HET)) %&gt;%
  dplyr::left_join(., final_founder_samples)
XY_hetrate_plot &lt;- ggplot(XY_hetrate, mapping = aes(x = predicted.sex, y = het_rate, fill = bg, label = sample)) + 
  geom_jitter(shape = 21, width = 0.1) + 
  facet_grid(.~Chromosome)
ggplotly(XY_hetrate_plot)</code></pre>
<div id="htmlwidget-7cb8309ee7030be5ebdb" style="width:672px;height:480px;" class="plotly html-widget"></div>
<script type="application/json" data-for="htmlwidget-7cb8309ee7030be5ebdb">{"x":{"data":[{"x":[0.982488580513746,0.924875468667597,1.05424942928366,1.08427409478463,0.915912109659985,1.01561492709443,0.964201006572694,0.900144240725785,1.00652572019026,1.05896316659637,0.94094023713842,1.06152193499729,1.05224550175481,0.978981118369848,0.904555687122047,1.07077294937335,0.999090785719454,0.996136801596731,1.02452094750479,0.90482233138755,0.930999382259324,1.08446688214317,1.09605944864452,1.09553925185464,1.02569215837866,1.9564020421356,1.9464507278055,2.03673815615475,1.96614583735354,2.02744652377442,2.02523711286485,2.09454828575254,2.03406706703827,1.94910292713903,2.00404319413938],"y":[0.543048159879717,0.575129563931489,0.304444706115232,0.301310264783902,0.423990344861616,0.46505265757592,0.624609380627822,0.244495759631197,0.297085880972445,0.561265260732234,0.282289171089428,0.473294166674593,0.515309336813863,0.283416428317626,0.209522162322519,0.203732391213528,0.249726620302206,0.439867821111146,0.277995454865799,0.236522835403154,0.444501337483525,0.403722794077011,0.0421692222183702,0.280146660943224,0.300930837046995,0.0386080226238177,0.0340162954114453,0.0529929220386677,0.0487429506497251,0.0286948402828038,0.0254865122462313,0.0276116461649236,0.0309026872483317,0.0427044231346761,0.0299434767990853],"text":["predicted.sex: f<br />het_rate: 0.54343434<br />bg: CROSS<br />sample: 129xCASTf005 (f)","predicted.sex: f<br />het_rate: 0.57575758<br />bg: CROSS<br />sample: 129xPWK 040F (f)","predicted.sex: f<br />het_rate: 0.30505051<br />bg: CROSS<br />sample: AB0095 (f)","predicted.sex: f<br />het_rate: 0.30101010<br />bg: CROSS<br />sample: AE-F (f)","predicted.sex: f<br />het_rate: 0.42424242<br />bg: CROSS<br />sample: AH0030 (f)","predicted.sex: f<br />het_rate: 0.46464646<br />bg: CROSS<br />sample: AJxPWKf001 (f)","predicted.sex: f<br />het_rate: 0.62424242<br />bg: CROSS<br />sample: B6xPWKf002 (f)","predicted.sex: f<br />het_rate: 0.24444444<br />bg: CROSS<br />sample: BD0018 (f)","predicted.sex: f<br />het_rate: 0.29696970<br />bg: CROSS<br />sample: BE0016 (f)","predicted.sex: f<br />het_rate: 0.56161616<br />bg: CROSS<br />sample: BF0016 (f)","predicted.sex: f<br />het_rate: 0.28282828<br />bg: CROSS<br />sample: CA0056 (f)","predicted.sex: f<br />het_rate: 0.47272727<br />bg: CROSS<br />sample: CASTxAJf001 (f)","predicted.sex: f<br />het_rate: 0.51515152<br />bg: CROSS<br />sample: CASTxNODf007 (f)","predicted.sex: f<br />het_rate: 0.28282828<br />bg: CROSS<br />sample: CASTxWSBf015 (f)","predicted.sex: f<br />het_rate: 0.21010101<br />bg: CROSS<br />sample: CB15916F (f)","predicted.sex: f<br />het_rate: 0.20404040<br />bg: CROSS<br />sample: CD0063 (f)","predicted.sex: f<br />het_rate: 0.25050505<br />bg: CROSS<br />sample: CE0005 (f)","predicted.sex: f<br />het_rate: 0.44040404<br />bg: CROSS<br />sample: CH0013 (f)","predicted.sex: f<br />het_rate: 0.27878788<br />bg: CROSS<br />sample: DA0018 (f)","predicted.sex: f<br />het_rate: 0.23636364<br />bg: CROSS<br />sample: DE0001 (f)","predicted.sex: f<br />het_rate: 0.44444444<br />bg: CROSS<br />sample: DH0141 (f)","predicted.sex: f<br />het_rate: 0.40404040<br />bg: CROSS<br />sample: EH15876F (f)","predicted.sex: f<br />het_rate: 0.04242424<br />bg: CROSS<br />sample: NODxPWKm004 (f)","predicted.sex: f<br />het_rate: 0.28080808<br />bg: CROSS<br />sample: WSBxCASTf003 (f)","predicted.sex: f<br />het_rate: 0.30101010<br />bg: CROSS<br />sample: WSBxPWKf003 (f)","predicted.sex: m<br />het_rate: 0.03838384<br />bg: CROSS<br />sample: 129xPWK037m (m)","predicted.sex: m<br />het_rate: 0.03434343<br />bg: CROSS<br />sample: AJxPWKm006 (m)","predicted.sex: m<br />het_rate: 0.05252525<br />bg: CROSS<br />sample: B6xPWKm006 (m)","predicted.sex: m<br />het_rate: 0.04848485<br />bg: CROSS<br />sample: B6xWSB 15713 (m)","predicted.sex: m<br />het_rate: 0.02828283<br />bg: CROSS<br />sample: CASTx129m001 (m)","predicted.sex: m<br />het_rate: 0.02626263<br />bg: CROSS<br />sample: CASTxAJm005 (m)","predicted.sex: m<br />het_rate: 0.02828283<br />bg: CROSS<br />sample: CASTxPWKm021 (m)","predicted.sex: m<br />het_rate: 0.03030303<br />bg: CROSS<br />sample: CASTxWSBm001 (m)","predicted.sex: m<br />het_rate: 0.04242424<br />bg: CROSS<br />sample: NODxPWKm003 (m)","predicted.sex: m<br />het_rate: 0.03030303<br />bg: CROSS<br />sample: WSBxPWKm001 (m)"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(248,118,109,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","name":"CROSS","legendgroup":"CROSS","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[0.935005040280521,0.911239273007959,1.09382997504435,1.08326230351813,1.0117499683518,0.974678373849019,0.91434909934178,1.03504212712869,1.01461309455335,1.08688580342568,1.08705368894152,0.941994646703824,1.06035777884535,1.07587991734035,0.905575673561543,0.903686098288745,1.01617548125796,1.0564899393823,1.0973600607831,1.09720108080655,0.996009663445875,1.04741789139807,1.04109721886925,0.950054372102022,0.923305060248822,2.01342866779305,1.91181238247082,1.90134713668376,2.04903583331034,2.0031762925908,1.95007699313574,1.98708034520969,1.98695435058326,1.95544145856984,2.08755388120189],"y":[0.19964622413796,0.199228436308678,0.20005208467519,0.200420976686056,0.199780085796509,0.20025895756754,0.200414905930785,0.200712795799052,0.200731920417812,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,0.199538975809936,0.199537043709571,0.199574243664892,0.200578467748081,0.200587794511682,0.199766578599811,0.199789591600573,null,null,null],"text":["predicted.sex: f<br />het_rate: 0.20000000<br />bg: CROSS<br />sample: 129xPWK 040F (f)","predicted.sex: f<br />het_rate: 0.20000000<br />bg: CROSS<br />sample: AE-F (f)","predicted.sex: f<br />het_rate: 0.20000000<br />bg: CROSS<br />sample: CASTxAJf001 (f)","predicted.sex: f<br />het_rate: 0.20000000<br />bg: CROSS<br />sample: CASTxNODf007 (f)","predicted.sex: f<br />het_rate: 0.20000000<br />bg: CROSS<br />sample: CASTxWSBf015 (f)","predicted.sex: f<br />het_rate: 0.20000000<br />bg: CROSS<br />sample: EH15876F (f)","predicted.sex: f<br />het_rate: 0.20000000<br />bg: CROSS<br />sample: NODxPWKm004 (f)","predicted.sex: f<br />het_rate: 0.20000000<br />bg: CROSS<br />sample: WSBxCASTf003 (f)","predicted.sex: f<br />het_rate: 0.20000000<br />bg: CROSS<br />sample: WSBxPWKf003 (f)","predicted.sex: f<br />het_rate:         NA<br />bg: CROSS<br />sample: 129xCASTf005 (f)","predicted.sex: f<br />het_rate:         NA<br />bg: CROSS<br />sample: AB0095 (f)","predicted.sex: f<br />het_rate:         NA<br />bg: CROSS<br />sample: AH0030 (f)","predicted.sex: f<br />het_rate:         NA<br />bg: CROSS<br />sample: AJxPWKf001 (f)","predicted.sex: f<br />het_rate:         NA<br />bg: CROSS<br />sample: B6xPWKf002 (f)","predicted.sex: f<br />het_rate:         NA<br />bg: CROSS<br />sample: BD0018 (f)","predicted.sex: f<br />het_rate:         NA<br />bg: CROSS<br />sample: BE0016 (f)","predicted.sex: f<br />het_rate:         NA<br />bg: CROSS<br />sample: BF0016 (f)","predicted.sex: f<br />het_rate:         NA<br />bg: CROSS<br />sample: CA0056 (f)","predicted.sex: f<br />het_rate:         NA<br />bg: CROSS<br />sample: CB15916F (f)","predicted.sex: f<br />het_rate:         NA<br />bg: CROSS<br />sample: CD0063 (f)","predicted.sex: f<br />het_rate:         NA<br />bg: CROSS<br />sample: CE0005 (f)","predicted.sex: f<br />het_rate:         NA<br />bg: CROSS<br />sample: CH0013 (f)","predicted.sex: f<br />het_rate:         NA<br />bg: CROSS<br />sample: DA0018 (f)","predicted.sex: f<br />het_rate:         NA<br />bg: CROSS<br />sample: DE0001 (f)","predicted.sex: f<br />het_rate:         NA<br />bg: CROSS<br />sample: DH0141 (f)","predicted.sex: m<br />het_rate: 0.20000000<br />bg: CROSS<br />sample: 129xPWK037m (m)","predicted.sex: m<br />het_rate: 0.20000000<br />bg: CROSS<br />sample: AJxPWKm006 (m)","predicted.sex: m<br />het_rate: 0.20000000<br />bg: CROSS<br />sample: B6xPWKm006 (m)","predicted.sex: m<br />het_rate: 0.20000000<br />bg: CROSS<br />sample: B6xWSB 15713 (m)","predicted.sex: m<br />het_rate: 0.20000000<br />bg: CROSS<br />sample: CASTxWSBm001 (m)","predicted.sex: m<br />het_rate: 0.20000000<br />bg: CROSS<br />sample: NODxPWKm003 (m)","predicted.sex: m<br />het_rate: 0.20000000<br />bg: CROSS<br />sample: WSBxPWKm001 (m)","predicted.sex: m<br />het_rate:         NA<br />bg: CROSS<br />sample: CASTx129m001 (m)","predicted.sex: m<br />het_rate:         NA<br />bg: CROSS<br />sample: CASTxAJm005 (m)","predicted.sex: m<br />het_rate:         NA<br />bg: CROSS<br />sample: CASTxPWKm021 (m)"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(248,118,109,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","name":"CROSS","legendgroup":"CROSS","showlegend":false,"xaxis":"x2","yaxis":"y","hoverinfo":"text","frame":null},{"x":[1.01705482597463,1.05924794469029,1.08699683495797,0.987069069640711,0.938010707264766,1.08252813853323,0.9894609129522,0.963547943113372,2.01793307331391,1.96946206321008,2.06128285177983,2.09304038644768,1.96243325863034],"y":[0.0344680788341646,0.0305579730845762,0.0511327740399524,0.0269873509183526,0.0337019615590271,0.0323621670149191,0.0223370625676982,0.0307973859293593,0.0378355067102897,0.0337601243237955,0.0279880958979931,0.040351969273855,0.0259071050588079],"text":["predicted.sex: f<br />het_rate: 0.03434343<br />bg: INBRED<br />sample: 129S1/SvImJ (f)","predicted.sex: f<br />het_rate: 0.03030303<br />bg: INBRED<br />sample: A/J (f)","predicted.sex: f<br />het_rate: 0.05050505<br />bg: INBRED<br />sample: C57BL/6J (f)","predicted.sex: f<br />het_rate: 0.02626263<br />bg: INBRED<br />sample: CAST/EiJ (f)","predicted.sex: f<br />het_rate: 0.03434343<br />bg: INBRED<br />sample: NOD/LtJ (f)","predicted.sex: f<br />het_rate: 0.03232323<br />bg: INBRED<br />sample: NZO/HlLtJ (f)","predicted.sex: f<br />het_rate: 0.02222222<br />bg: INBRED<br />sample: PWK/PhJ (f)","predicted.sex: f<br />het_rate: 0.03030303<br />bg: INBRED<br />sample: WSB/EiJ (f)","predicted.sex: m<br />het_rate: 0.03838384<br />bg: INBRED<br />sample: 129S1/SvImJm212 (m)","predicted.sex: m<br />het_rate: 0.03434343<br />bg: INBRED<br />sample: A/Jm111 (m)","predicted.sex: m<br />het_rate: 0.02828283<br />bg: INBRED<br />sample: CAST/EiJm42 (m)","predicted.sex: m<br />het_rate: 0.04040404<br />bg: INBRED<br />sample: NZO/HlLtJm51 (m)","predicted.sex: m<br />het_rate: 0.02626263<br />bg: INBRED<br />sample: PWK/PhJm175 (m)"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(0,191,196,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","name":"INBRED","legendgroup":"INBRED","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[1.01844026385806,0.99677460193634,1.05217810757458,1.04649313758127,0.920926824491471,1.02875259891152,1.06424281988293,0.97891857502982,1.97715891920961,2.07267889636569,2.04022400206886,2.03132623247802,2.06198041676544],"y":[0.200068551784739,0.200470832400024,0.199343190387719,0.199861326273343,0.200081853899315,0.39920864303681,null,null,0.200651763917671,0.200009734865767,0.200075745911836,null,null],"text":["predicted.sex: f<br />het_rate: 0.20000000<br />bg: INBRED<br />sample: 129S1/SvImJ (f)","predicted.sex: f<br />het_rate: 0.20000000<br />bg: INBRED<br />sample: A/J (f)","predicted.sex: f<br />het_rate: 0.20000000<br />bg: INBRED<br />sample: C57BL/6J (f)","predicted.sex: f<br />het_rate: 0.20000000<br />bg: INBRED<br />sample: CAST/EiJ (f)","predicted.sex: f<br />het_rate: 0.20000000<br />bg: INBRED<br />sample: NOD/LtJ (f)","predicted.sex: f<br />het_rate: 0.40000000<br />bg: INBRED<br />sample: WSB/EiJ (f)","predicted.sex: f<br />het_rate:         NA<br />bg: INBRED<br />sample: NZO/HlLtJ (f)","predicted.sex: f<br />het_rate:         NA<br />bg: INBRED<br />sample: PWK/PhJ (f)","predicted.sex: m<br />het_rate: 0.20000000<br />bg: INBRED<br />sample: 129S1/SvImJm212 (m)","predicted.sex: m<br />het_rate: 0.20000000<br />bg: INBRED<br />sample: A/Jm111 (m)","predicted.sex: m<br />het_rate: 0.20000000<br />bg: INBRED<br />sample: NZO/HlLtJm51 (m)","predicted.sex: m<br />het_rate:         NA<br />bg: INBRED<br />sample: CAST/EiJm42 (m)","predicted.sex: m<br />het_rate:         NA<br />bg: INBRED<br />sample: PWK/PhJm175 (m)"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(0,191,196,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","name":"INBRED","legendgroup":"INBRED","showlegend":false,"xaxis":"x2","yaxis":"y","hoverinfo":"text","frame":null},{"x":[1.96776085877791,1.92639163425192,1.98703916668892,2.09287815266289,2.08545185769908,2.04035494946875],"y":[0.0389196263987458,0.0349869222818601,0.0457430006405621,0.0410906690050556,0.0363232788838672,0.0443928338282488],"text":["predicted.sex: m<br />het_rate: 0.03838384<br />bg: NA<br />sample: #1-m-9376 (m)","predicted.sex: m<br />het_rate: 0.03434343<br />bg: NA<br />sample: #2-m-9376 (m)","predicted.sex: m<br />het_rate: 0.04646465<br />bg: NA<br />sample: #3-m-9376 (m)","predicted.sex: m<br />het_rate: 0.04040404<br />bg: NA<br />sample: #4-m-9376 (m)","predicted.sex: m<br />het_rate: 0.03636364<br />bg: NA<br />sample: #5-m-9376 (m)","predicted.sex: m<br />het_rate: 0.04444444<br />bg: NA<br />sample: #6-m-9376 (m)"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(127,127,127,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","name":"NA","legendgroup":"NA","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[2.08829595530406,2.07436471981928,2.07199621838517,1.92317324578762,2.00418671900406,2.09666753220372],"y":[0.199611361127234,0.200161511787787,0.199237555865989,0.199522943868104,0.200500973273784,null],"text":["predicted.sex: m<br />het_rate: 0.20000000<br />bg: NA<br />sample: #1-m-9376 (m)","predicted.sex: m<br />het_rate: 0.20000000<br />bg: NA<br />sample: #2-m-9376 (m)","predicted.sex: m<br />het_rate: 0.20000000<br />bg: NA<br />sample: #3-m-9376 (m)","predicted.sex: m<br />het_rate: 0.20000000<br />bg: NA<br />sample: #5-m-9376 (m)","predicted.sex: m<br />het_rate: 0.20000000<br />bg: NA<br />sample: #6-m-9376 (m)","predicted.sex: m<br />het_rate:         NA<br />bg: NA<br />sample: #4-m-9376 (m)"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(127,127,127,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","name":"NA","legendgroup":"NA","showlegend":false,"xaxis":"x2","yaxis":"y","hoverinfo":"text","frame":null}],"layout":{"margin":{"t":37.9178082191781,"r":18.9954337899543,"b":40.1826484018265,"l":43.1050228310502},"plot_bgcolor":"rgba(235,235,235,1)","paper_bgcolor":"rgba(255,255,255,1)","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187},"xaxis":{"domain":[0,0.489128071319852],"automargin":true,"type":"linear","autorange":false,"range":[0.4,2.6],"tickmode":"array","ticktext":["f","m"],"tickvals":[1,2],"categoryorder":"array","categoryarray":["f","m"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.65296803652968,"tickwidth":0.66417600664176,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":11.689497716895},"tickangle":-0,"showline":false,"linecolor":null,"linewidth":0,"showgrid":true,"gridcolor":"rgba(255,255,255,1)","gridwidth":0.66417600664176,"zeroline":false,"anchor":"y","title":"","hoverformat":".2f"},"annotations":[{"text":"predicted.sex","x":0.5,"y":0,"showarrow":false,"ax":0,"ay":0,"font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187},"xref":"paper","yref":"paper","textangle":-0,"xanchor":"center","yanchor":"top","annotationType":"axis","yshift":-21.9178082191781},{"text":"het_rate","x":0,"y":0.5,"showarrow":false,"ax":0,"ay":0,"font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187},"xref":"paper","yref":"paper","textangle":-90,"xanchor":"right","yanchor":"center","annotationType":"axis","xshift":-27.7625570776256},{"text":"X","x":0.244564035659926,"y":1,"showarrow":false,"ax":0,"ay":0,"font":{"color":"rgba(26,26,26,1)","family":"","size":11.689497716895},"xref":"paper","yref":"paper","textangle":-0,"xanchor":"center","yanchor":"bottom"},{"text":"Y","x":0.755435964340074,"y":1,"showarrow":false,"ax":0,"ay":0,"font":{"color":"rgba(26,26,26,1)","family":"","size":11.689497716895},"xref":"paper","yref":"paper","textangle":-0,"xanchor":"center","yanchor":"bottom"}],"yaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[-0.00777655333530798,0.654722996530829],"tickmode":"array","ticktext":["0.0","0.2","0.4","0.6"],"tickvals":[0,0.2,0.4,0.6],"categoryorder":"array","categoryarray":["0.0","0.2","0.4","0.6"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.65296803652968,"tickwidth":0.66417600664176,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":11.689497716895},"tickangle":-0,"showline":false,"linecolor":null,"linewidth":0,"showgrid":true,"gridcolor":"rgba(255,255,255,1)","gridwidth":0.66417600664176,"zeroline":false,"anchor":"x","title":"","hoverformat":".2f"},"shapes":[{"type":"rect","fillcolor":null,"line":{"color":null,"width":0,"linetype":[]},"yref":"paper","xref":"paper","x0":0,"x1":0.489128071319852,"y0":0,"y1":1},{"type":"rect","fillcolor":"rgba(217,217,217,1)","line":{"color":"transparent","width":0.66417600664176,"linetype":"solid"},"yref":"paper","xref":"paper","x0":0,"x1":0.489128071319852,"y0":0,"y1":23.37899543379,"yanchor":1,"ysizemode":"pixel"},{"type":"rect","fillcolor":null,"line":{"color":null,"width":0,"linetype":[]},"yref":"paper","xref":"paper","x0":0.510871928680148,"x1":1,"y0":0,"y1":1},{"type":"rect","fillcolor":"rgba(217,217,217,1)","line":{"color":"transparent","width":0.66417600664176,"linetype":"solid"},"yref":"paper","xref":"paper","x0":0.510871928680148,"x1":1,"y0":0,"y1":23.37899543379,"yanchor":1,"ysizemode":"pixel"}],"xaxis2":{"type":"linear","autorange":false,"range":[0.4,2.6],"tickmode":"array","ticktext":["f","m"],"tickvals":[1,2],"categoryorder":"array","categoryarray":["f","m"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.65296803652968,"tickwidth":0.66417600664176,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":11.689497716895},"tickangle":-0,"showline":false,"linecolor":null,"linewidth":0,"showgrid":true,"domain":[0.510871928680148,1],"gridcolor":"rgba(255,255,255,1)","gridwidth":0.66417600664176,"zeroline":false,"anchor":"y","title":"","hoverformat":".2f"},"showlegend":true,"legend":{"bgcolor":"rgba(255,255,255,1)","bordercolor":"transparent","borderwidth":1.88976377952756,"font":{"color":"rgba(0,0,0,1)","family":"","size":11.689497716895},"title":{"text":"bg","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187}}},"hovermode":"closest","barmode":"relative"},"config":{"doubleClick":"reset","modeBarButtonsToAdd":["hoverclosest","hovercompare"],"showSendToCloud":false},"source":"A","attrs":{"80cc42e78dfe":{"x":{},"y":{},"fill":{},"label":{},"type":"scatter"}},"cur_data":"80cc42e78dfe","visdat":{"80cc42e78dfe":["function (y) ","x"]},"highlight":{"on":"plotly_click","persistent":false,"dynamic":false,"selectize":false,"opacityDim":0.2,"selected":{"opacity":1},"debounce":0},"shinyEvents":["plotly_hover","plotly_click","plotly_selected","plotly_relayout","plotly_brushed","plotly_brushing","plotly_clickannotation","plotly_doubleclick","plotly_deselect","plotly_afterplot","plotly_sunburstclick"],"base_url":"https://plot.ly"},"evals":[],"jsHooks":[]}</script>
<pre class="r"><code># Input: Sex chromosome probe intensities for each marker with 1) marker metdata, 2) marker and sample flags, 3) background and sex predictions
Xchr.int &lt;- prelim.predicted.sexes %&gt;%
  dplyr::ungroup() %&gt;%
  dplyr::filter(marker_flag != &quot;FLAG&quot;,
                Chromosome == &quot;X&quot;) %&gt;%
  dplyr::mutate(x.chr.int = X + Y,
                theta = as.numeric(theta)) %&gt;%
  dplyr::group_by(sample, predicted.sex) %&gt;%
  dplyr::summarise(mean.x.chr.int = mean(x.chr.int),
                   mean.x = mean(X),
                   mean.theta = mean(theta))
# Expected output: Sample-averaged summed x- and y-channel probe intensities for all chromosome X markers. Note: replicated sample information collapses at this step. This is tolerable under the assumption that the samples with identical names are in fact duplicates of the same individual.

#   sample        predicted.sex mean.x.chr.int
#   &lt;chr&gt;         &lt;chr&gt;                  &lt;dbl&gt;
# 1 #1-m-9376 (m) m                       1.10
# 2 #2-m-9376 (m) m                       1.07
# 3 #3-m-9376 (m) m                       1.10
# 4 #4-m-9376 (m) m                       1.13
# 5 #5-m-9376 (m) m                       1.12
# 6 #6-m-9376 (m) m                       1.11


Ychr.int &lt;- prelim.predicted.sexes %&gt;%
  dplyr::ungroup() %&gt;%
  dplyr::mutate(theta = as.numeric(theta)) %&gt;%
  dplyr::filter(marker_flag != &quot;FLAG&quot;,
                Chromosome == &quot;Y&quot;) %&gt;%
  dplyr::group_by(sample, predicted.sex) %&gt;%
  dplyr::summarise(mean.y.int = mean(Y),
                   mean.theta = mean(theta))</code></pre>
<pre><code>Warning in mask$eval_all_mutate(quo): NAs introduced by coercion</code></pre>
<pre class="r"><code># Expected output: Sample-averaged y-channel probe intensities for all chromosome Y markers. Note: replicated sample information collapses at this step. This is tolerable under the assumption that the samples with identical names are in fact duplicates of the same individual.

#   sample        predicted.sex mean.y.int
#   &lt;chr&gt;         &lt;chr&gt;              &lt;dbl&gt;
# 1 #1-m-9376 (m) m                  0.826
# 2 #2-m-9376 (m) m                  0.804
# 3 #3-m-9376 (m) m                  0.729
# 4 #4-m-9376 (m) m                  0.804
# 5 #5-m-9376 (m) m                  0.826
# 6 #6-m-9376 (m) m                  0.733


# Column binding the two intensities if the sample information matches
if(unique(Xchr.int$sample == Ychr.int$sample) == TRUE){
  sex.chr.intensities &lt;- cbind(Xchr.int,Ychr.int$mean.y.int)
  colnames(sex.chr.intensities) &lt;- c(&quot;sample&quot;,&quot;predicted.sex&quot;,&quot;sumxy_int&quot;,&quot;x_int&quot;,&quot;mean_theta&quot;,&quot;y_int&quot;)
}

# Interactive visualization of sex prediction results. Sample are colored according to predicted sex. &quot;Unknown&quot; samples are plotted black, and flagged/bad samples are triangles.
predicted.sex.plot.palettes &lt;- sex.chr.intensities %&gt;%
  dplyr::ungroup() %&gt;%
  dplyr::distinct(sample, predicted.sex) %&gt;%
  dplyr::mutate(predicted.sex.palette = dplyr::case_when(predicted.sex == &quot;f&quot; ~ &quot;#5856b7&quot;,
                                                         predicted.sex == &quot;m&quot; ~ &quot;#eeb868&quot;,
                                                         predicted.sex == &quot;unknown&quot; ~ &quot;black&quot;))
predicted.sex.palette &lt;- predicted.sex.plot.palettes$predicted.sex.palette
names(predicted.sex.palette) &lt;- predicted.sex.plot.palettes$predicted.sex
mean.x.intensities.by.sex.plot &lt;- ggplot(sex.chr.intensities, 
                                         mapping = aes(x = sumxy_int, 
                                                       y = y_int, 
                                                       colour = predicted.sex,
                                                       text = sample)) + 
  geom_point() + 
  scale_colour_manual(values = predicted.sex.palette) + 
  # facet_grid(.~chr) +
  QCtheme
ggplotly(mean.x.intensities.by.sex.plot, 
         tooltip = c(&quot;text&quot;,&quot;label&quot;))</code></pre>
<div id="htmlwidget-4fbe5ee9c104401b156b" style="width:672px;height:480px;" class="plotly html-widget"></div>
<script type="application/json" data-for="htmlwidget-4fbe5ee9c104401b156b">{"x":{"data":[{"x":[1.22752235772358,1.39236178861789,1.37565853658537,1.1997256097561,1.31170731707317,1.30744512195122,1.42873780487805,1.36005487804878,1.37754471544715,1.30004268292683,1.29311382113821,1.40788414634146,1.05548170731707,1.32752032520325,1.16435569105691,1.38215447154472,1.41708943089431,1.29078861788618,1.29383333333333,1.31618292682927,1.42118292682927,1.36312398373984,1.35930691056911,1.45293089430894,1.42187601626016,1.34756504065041,1.26053861788618,1.12823170731707,1.17944105691057,1.12339634146341,1.03208130081301,1.29611788617886,1.29557926829268],"y":[0.875666666666667,0.915333333333333,0.852,0.876333333333333,0.868333333333333,0.914666666666667,0.829666666666667,0.861,0.808,0.880333333333333,0.824333333333333,0.765333333333333,0.728333333333333,0.878666666666667,0.858333333333333,0.901333333333333,0.936666666666667,0.889,0.822666666666667,0.910333333333333,0.853333333333333,0.862333333333333,0.887,0.946,0.945,0.856666666666667,0.904333333333333,0.839333333333333,0.850333333333333,0.719333333333333,0.724666666666667,0.879333333333333,0.841666666666667],"text":["129S1/SvImJ (f)","129xCASTf005 (f)","129xPWK 040F (f)","A/J (f)","AB0095 (f)","AE-F (f)","AH0030 (f)","AJxPWKf001 (f)","B6xPWKf002 (f)","BD0018 (f)","BE0016 (f)","BF0016 (f)","C57BL/6J (f)","CA0056 (f)","CAST/EiJ (f)","CASTxAJf001 (f)","CASTxNODf007 (f)","CASTxWSBf015 (f)","CB15916F (f)","CD0063 (f)","CE0005 (f)","CH0013 (f)","DA0018 (f)","DE0001 (f)","DH0141 (f)","EH15876F (f)","NOD/LtJ (f)","NODxPWKm004 (f)","NZO/HlLtJ (f)","PWK/PhJ (f)","WSB/EiJ (f)","WSBxCASTf003 (f)","WSBxPWKf003 (f)"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(88,86,183,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(88,86,183,1)"}},"hoveron":"points","name":"f","legendgroup":"f","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[1.09527235772358,1.06566056910569,1.10097764227642,1.13062195121951,1.11763414634146,1.11185162601626,1.10681910569106,1.11593089430894,1.04588211382114,1.10451829268293,1.12718699186992,1.04722357723577,1.04555691056911,1.08422154471545,1.07257317073171,1.07609146341463,1.06887398373984,1.1287418699187,1.08084959349594,1.02055487804878,1.09832723577236],"y":[0.826,0.804,0.729333333333333,0.803666666666667,0.826333333333333,0.733333333333333,0.778666666666667,0.797,0.788,0.829666666666667,0.743,0.729666666666667,0.706333333333333,0.771333333333333,0.741,0.761,0.759333333333333,0.843,0.784333333333333,0.627333333333333,0.808666666666667],"text":["#1-m-9376 (m)","#2-m-9376 (m)","#3-m-9376 (m)","#4-m-9376 (m)","#5-m-9376 (m)","#6-m-9376 (m)","129S1/SvImJm212 (m)","129xPWK037m (m)","A/Jm111 (m)","AJxPWKm006 (m)","B6xPWKm006 (m)","B6xWSB 15713 (m)","CAST/EiJm42 (m)","CASTx129m001 (m)","CASTxAJm005 (m)","CASTxPWKm021 (m)","CASTxWSBm001 (m)","NODxPWKm003 (m)","NZO/HlLtJm51 (m)","PWK/PhJm175 (m)","WSBxPWKm001 (m)"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(238,184,104,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(238,184,104,1)"}},"hoveron":"points","name":"m","legendgroup":"m","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null}],"layout":{"margin":{"t":26.2283105022831,"r":7.30593607305936,"b":40.1826484018265,"l":43.1050228310502},"plot_bgcolor":"rgba(255,255,255,1)","paper_bgcolor":"rgba(255,255,255,1)","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187},"xaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[0.998936077235772,1.47454969512195],"tickmode":"array","ticktext":["1.0","1.1","1.2","1.3","1.4"],"tickvals":[1,1.1,1.2,1.3,1.4],"categoryorder":"array","categoryarray":["1.0","1.1","1.2","1.3","1.4"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.65296803652968,"tickwidth":0.66417600664176,"showticklabels":true,"tickfont":{"color":"rgba(0,0,0,1)","family":"","size":11.689497716895},"tickangle":-0,"showline":false,"linecolor":null,"linewidth":0,"showgrid":false,"gridcolor":null,"gridwidth":0,"zeroline":false,"anchor":"y","title":{"text":"sumxy_int","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187}},"hoverformat":".2f"},"yaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[0.6114,0.961933333333333],"tickmode":"array","ticktext":["0.7","0.8","0.9"],"tickvals":[0.7,0.8,0.9],"categoryorder":"array","categoryarray":["0.7","0.8","0.9"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.65296803652968,"tickwidth":0.66417600664176,"showticklabels":true,"tickfont":{"color":"rgba(0,0,0,1)","family":"","size":11.689497716895},"tickangle":-0,"showline":false,"linecolor":null,"linewidth":0,"showgrid":false,"gridcolor":null,"gridwidth":0,"zeroline":false,"anchor":"x","title":{"text":"y_int","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187}},"hoverformat":".2f"},"shapes":[{"type":"rect","fillcolor":"transparent","line":{"color":"rgba(51,51,51,1)","width":0.66417600664176,"linetype":"solid"},"yref":"paper","xref":"paper","x0":0,"x1":1,"y0":0,"y1":1}],"showlegend":true,"legend":{"bgcolor":"rgba(255,255,255,1)","bordercolor":"transparent","borderwidth":1.88976377952756,"font":{"color":"rgba(0,0,0,1)","family":"","size":11.689497716895},"title":{"text":"predicted.sex","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187}}},"hovermode":"closest","barmode":"relative"},"config":{"doubleClick":"reset","modeBarButtonsToAdd":["hoverclosest","hovercompare"],"showSendToCloud":false},"source":"A","attrs":{"80cc63eac1cd":{"x":{},"y":{},"colour":{},"text":{},"type":"scatter"}},"cur_data":"80cc63eac1cd","visdat":{"80cc63eac1cd":["function (y) ","x"]},"highlight":{"on":"plotly_click","persistent":false,"dynamic":false,"selectize":false,"opacityDim":0.2,"selected":{"opacity":1},"debounce":0},"shinyEvents":["plotly_hover","plotly_click","plotly_selected","plotly_relayout","plotly_brushed","plotly_brushing","plotly_clickannotation","plotly_doubleclick","plotly_deselect","plotly_afterplot","plotly_sunburstclick"],"base_url":"https://plot.ly"},"evals":[],"jsHooks":[]}</script>
<p>Combining plots of heterozygosity rates of X and Y chromosomes and
the sex chromosome probe intensities, we observed four samples that were
originally classified as females that were likely male samples,
including one named “NODxPWKm004 (f)” displaying proper sample
nomenclature for a male sample. We have manually assigned these samples
the appropriate sex designation.</p>
<pre class="r"><code>final.predicted.sex &lt;- prelim.predicted.sexes %&gt;%
  dplyr::mutate(inferred.sex = if_else(condition = sample %in% c(&quot;PWK/PhJ (f)&quot;,
                                                                 &quot;NODxPWKm004 (f)&quot;,
                                                                 &quot;C57BL/6J (f)&quot;,
                                                                 &quot;WSB/EiJ (f)&quot;), 
                                       true = &quot;m&quot;, 
                                       false = predicted.sex),
                resexed = if_else(condition = sample %in% c(&quot;PWK/PhJ (f)&quot;,
                                                                 &quot;NODxPWKm004 (f)&quot;,
                                                                 &quot;C57BL/6J (f)&quot;,
                                                                 &quot;WSB/EiJ (f)&quot;), 
                                       true = T, 
                                       false = F))


## Taking the first marker as a sample and tabulating the number of samples for each predicted sex
predicted.sex.table &lt;- final.predicted.sex %&gt;%
  dplyr::filter(marker %in% unique(prelim.predicted.sexes$marker)[1]) %&gt;%
  dplyr::select(sample, inferred.sex) %&gt;%
  dplyr::group_by(inferred.sex) %&gt;%
  dplyr::count()</code></pre>
<p>This captured 29 female samples, 25 male samples, leaving samples of
unknown predicted sex from nomenclature alone.</p>
<pre class="r"><code>Xchr.int &lt;- final.predicted.sex %&gt;%
  dplyr::ungroup() %&gt;%
  dplyr::filter(marker_flag != &quot;FLAG&quot;,
                Chromosome == &quot;X&quot;) %&gt;%
  dplyr::mutate(x.chr.int = X + Y,
                theta = as.numeric(theta)) %&gt;%
  dplyr::group_by(sample, inferred.sex) %&gt;%
  dplyr::summarise(mean.x.chr.int = mean(x.chr.int),
                   mean.x = mean(X),
                   mean.theta = mean(theta))


Ychr.int &lt;- final.predicted.sex %&gt;%
  dplyr::ungroup() %&gt;%
  dplyr::mutate(theta = as.numeric(theta)) %&gt;%
  dplyr::filter(marker_flag != &quot;FLAG&quot;,
                Chromosome == &quot;Y&quot;) %&gt;%
  dplyr::group_by(sample, inferred.sex) %&gt;%
  dplyr::summarise(mean.y.int = mean(Y),
                   mean.theta = mean(theta))</code></pre>
<pre><code>Warning in mask$eval_all_mutate(quo): NAs introduced by coercion</code></pre>
<pre class="r"><code># Column binding the two intensities if the sample information matches
if(unique(Xchr.int$sample == Ychr.int$sample) == TRUE){
  sex.chr.intensities &lt;- cbind(Xchr.int,Ychr.int$mean.y.int)
  colnames(sex.chr.intensities) &lt;- c(&quot;sample&quot;,&quot;inferred.sex&quot;,&quot;sumxy_int&quot;,&quot;x_int&quot;,&quot;mean_theta&quot;,&quot;y_int&quot;)
}

# Interactive visualization of sex prediction results. Sample are colored according to predicted sex. &quot;Unknown&quot; samples are plotted black, and flagged/bad samples are triangles.
final.sex.plot.palettes &lt;- sex.chr.intensities %&gt;%
  dplyr::ungroup() %&gt;%
  dplyr::distinct(sample, inferred.sex) %&gt;%
  dplyr::mutate(final.sex.palette = dplyr::case_when(inferred.sex == &quot;f&quot; ~ &quot;#5856b7&quot;,
                                                     inferred.sex == &quot;m&quot; ~ &quot;#eeb868&quot;,
                                                     inferred.sex == &quot;unknown&quot; ~ &quot;black&quot;))
final.sex.palette &lt;- final.sex.plot.palettes$final.sex.palette
names(predicted.sex.palette) &lt;- final.sex.plot.palettes$inferred.sex
mean.x.intensities.by.sex.plot &lt;- ggplot(sex.chr.intensities %&gt;%
                                           dplyr::left_join(., final.predicted.sex %&gt;% 
                                                              dplyr::distinct(sample, inferred.sex, resexed)), 
                                         mapping = aes(x = sumxy_int, 
                                                       y = y_int, 
                                                       colour = inferred.sex,
                                                       fill = resexed,
                                                       text = sample)) + 
  geom_point() + 
  scale_colour_manual(values = predicted.sex.palette) + 
  scale_fill_manual(values = c(&quot;black&quot;,&quot;white&quot;)) +
  # facet_grid(.~chr) +
  QCtheme
ggplotly(mean.x.intensities.by.sex.plot, 
         tooltip = c(&quot;text&quot;,&quot;label&quot;))</code></pre>
<div id="htmlwidget-b91531a67c690b508a43" style="width:672px;height:480px;" class="plotly html-widget"></div>
<script type="application/json" data-for="htmlwidget-b91531a67c690b508a43">{"x":{"data":[{"x":[1.22752235772358,1.39236178861789,1.37565853658537,1.1997256097561,1.31170731707317,1.30744512195122,1.42873780487805,1.36005487804878,1.37754471544715,1.30004268292683,1.29311382113821,1.40788414634146,1.32752032520325,1.16435569105691,1.38215447154472,1.41708943089431,1.29078861788618,1.29383333333333,1.31618292682927,1.42118292682927,1.36312398373984,1.35930691056911,1.45293089430894,1.42187601626016,1.34756504065041,1.26053861788618,1.17944105691057,1.29611788617886,1.29557926829268],"y":[0.875666666666667,0.915333333333333,0.852,0.876333333333333,0.868333333333333,0.914666666666667,0.829666666666667,0.861,0.808,0.880333333333333,0.824333333333333,0.765333333333333,0.878666666666667,0.858333333333333,0.901333333333333,0.936666666666667,0.889,0.822666666666667,0.910333333333333,0.853333333333333,0.862333333333333,0.887,0.946,0.945,0.856666666666667,0.904333333333333,0.850333333333333,0.879333333333333,0.841666666666667],"text":["129S1/SvImJ (f)","129xCASTf005 (f)","129xPWK 040F (f)","A/J (f)","AB0095 (f)","AE-F (f)","AH0030 (f)","AJxPWKf001 (f)","B6xPWKf002 (f)","BD0018 (f)","BE0016 (f)","BF0016 (f)","CA0056 (f)","CAST/EiJ (f)","CASTxAJf001 (f)","CASTxNODf007 (f)","CASTxWSBf015 (f)","CB15916F (f)","CD0063 (f)","CE0005 (f)","CH0013 (f)","DA0018 (f)","DE0001 (f)","DH0141 (f)","EH15876F (f)","NOD/LtJ (f)","NZO/HlLtJ (f)","WSBxCASTf003 (f)","WSBxPWKf003 (f)"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(0,0,0,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(88,86,183,1)"}},"hoveron":"points","name":"(f,FALSE)","legendgroup":"(f,FALSE)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[1.09527235772358,1.06566056910569,1.10097764227642,1.13062195121951,1.11763414634146,1.11185162601626,1.10681910569106,1.11593089430894,1.04588211382114,1.10451829268293,1.12718699186992,1.04722357723577,1.04555691056911,1.08422154471545,1.07257317073171,1.07609146341463,1.06887398373984,1.1287418699187,1.08084959349594,1.02055487804878,1.09832723577236],"y":[0.826,0.804,0.729333333333333,0.803666666666667,0.826333333333333,0.733333333333333,0.778666666666667,0.797,0.788,0.829666666666667,0.743,0.729666666666667,0.706333333333333,0.771333333333333,0.741,0.761,0.759333333333333,0.843,0.784333333333333,0.627333333333333,0.808666666666667],"text":["#1-m-9376 (m)","#2-m-9376 (m)","#3-m-9376 (m)","#4-m-9376 (m)","#5-m-9376 (m)","#6-m-9376 (m)","129S1/SvImJm212 (m)","129xPWK037m (m)","A/Jm111 (m)","AJxPWKm006 (m)","B6xPWKm006 (m)","B6xWSB 15713 (m)","CAST/EiJm42 (m)","CASTx129m001 (m)","CASTxAJm005 (m)","CASTxPWKm021 (m)","CASTxWSBm001 (m)","NODxPWKm003 (m)","NZO/HlLtJm51 (m)","PWK/PhJm175 (m)","WSBxPWKm001 (m)"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(0,0,0,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(238,184,104,1)"}},"hoveron":"points","name":"(m,FALSE)","legendgroup":"(m,FALSE)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[1.05548170731707,1.12823170731707,1.12339634146341,1.03208130081301],"y":[0.728333333333333,0.839333333333333,0.719333333333333,0.724666666666667],"text":["C57BL/6J (f)","NODxPWKm004 (f)","PWK/PhJ (f)","WSB/EiJ (f)"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(255,255,255,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(238,184,104,1)"}},"hoveron":"points","name":"(m,TRUE)","legendgroup":"(m,TRUE)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null}],"layout":{"margin":{"t":26.2283105022831,"r":7.30593607305936,"b":40.1826484018265,"l":43.1050228310502},"plot_bgcolor":"rgba(255,255,255,1)","paper_bgcolor":"rgba(255,255,255,1)","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187},"xaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[0.998936077235772,1.47454969512195],"tickmode":"array","ticktext":["1.0","1.1","1.2","1.3","1.4"],"tickvals":[1,1.1,1.2,1.3,1.4],"categoryorder":"array","categoryarray":["1.0","1.1","1.2","1.3","1.4"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.65296803652968,"tickwidth":0.66417600664176,"showticklabels":true,"tickfont":{"color":"rgba(0,0,0,1)","family":"","size":11.689497716895},"tickangle":-0,"showline":false,"linecolor":null,"linewidth":0,"showgrid":false,"gridcolor":null,"gridwidth":0,"zeroline":false,"anchor":"y","title":{"text":"sumxy_int","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187}},"hoverformat":".2f"},"yaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[0.6114,0.961933333333333],"tickmode":"array","ticktext":["0.7","0.8","0.9"],"tickvals":[0.7,0.8,0.9],"categoryorder":"array","categoryarray":["0.7","0.8","0.9"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.65296803652968,"tickwidth":0.66417600664176,"showticklabels":true,"tickfont":{"color":"rgba(0,0,0,1)","family":"","size":11.689497716895},"tickangle":-0,"showline":false,"linecolor":null,"linewidth":0,"showgrid":false,"gridcolor":null,"gridwidth":0,"zeroline":false,"anchor":"x","title":{"text":"y_int","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187}},"hoverformat":".2f"},"shapes":[{"type":"rect","fillcolor":"transparent","line":{"color":"rgba(51,51,51,1)","width":0.66417600664176,"linetype":"solid"},"yref":"paper","xref":"paper","x0":0,"x1":1,"y0":0,"y1":1}],"showlegend":true,"legend":{"bgcolor":"rgba(255,255,255,1)","bordercolor":"transparent","borderwidth":1.88976377952756,"font":{"color":"rgba(0,0,0,1)","family":"","size":11.689497716895},"title":{"text":"inferred.sex<br />resexed","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187}}},"hovermode":"closest","barmode":"relative"},"config":{"doubleClick":"reset","modeBarButtonsToAdd":["hoverclosest","hovercompare"],"showSendToCloud":false},"source":"A","attrs":{"80cc4f387b2c":{"x":{},"y":{},"colour":{},"fill":{},"text":{},"type":"scatter"}},"cur_data":"80cc4f387b2c","visdat":{"80cc4f387b2c":["function (y) ","x"]},"highlight":{"on":"plotly_click","persistent":false,"dynamic":false,"selectize":false,"opacityDim":0.2,"selected":{"opacity":1},"debounce":0},"shinyEvents":["plotly_hover","plotly_click","plotly_selected","plotly_relayout","plotly_brushed","plotly_brushing","plotly_clickannotation","plotly_doubleclick","plotly_deselect","plotly_afterplot","plotly_sunburstclick"],"base_url":"https://plot.ly"},"evals":[],"jsHooks":[]}</script>
</div>
<div id="validating-reference-sample-genetic-backgrounds"
class="section level3">
<h3>Validating reference sample genetic backgrounds</h3>
<p>A key component of sample QC for our purposes is knowing that markers
that we expect to deliver the consensus genotype (<em>i.e.</em> in a
cross) actually provide us the correct strain information and allow us
to correctly infer haplotypes.</p>
<pre class="r"><code>sample.predicted.sexes.df &lt;- final.predicted.sex %&gt;%
  dplyr::distinct(sample, inferred.sex)

# Join all genotypes to founder-derived samples, filter away bad markers, and reduce down to unique rows for each sample and marker genotype
# Inputs: 
# 1) Founder sample metadata (colnames(all_founder_samples_parents) = &quot;sample&quot;        &quot;dam&quot;           &quot;sire&quot;          &quot;bad_sample&quot;    &quot;inferred.sex&quot;  &quot;resexed&quot;       &quot;bg&quot;            &quot;founder&quot;      &quot;weird.founder&quot;
# 2) All sample genotypes with flag information
founder_sample_genos &lt;- final_founder_samples %&gt;%
  dplyr::left_join(., sample.predicted.sexes.df) %&gt;%
  dplyr::select(-founder) %&gt;%
  dplyr::left_join(.,genos.flagged) %&gt;%
  dplyr::filter(marker_flag != &quot;FLAG&quot;,
                genotype != &quot;N&quot;) %&gt;%
  dplyr::distinct(sample, marker, genotype, inferred.sex, dam, sire)


# Creating a wide genotype table to compute the distance matrix for the dendrogram, filtering out the markers with multiple genotype calls per sample.
wide_founder_sample_genos &lt;- founder_sample_genos %&gt;%
  dplyr::select(sample, marker, genotype) %&gt;%
  tidyr::pivot_wider(names_from = marker, values_from = genotype)

# Genotype calls at this point are still in letter form (i.e. A, T, or H for het). In order to calculate the distance matrix, we had to recode each marker&#39;s genotype information into 0, 1 for hets or 2. This process is applied column-wise.
recoded_wide_sample_genos &lt;- suppressWarnings(data.frame(apply(wide_founder_sample_genos[,2:ncol(wide_founder_sample_genos)], 2, recodeCalls)))
dend.labels &lt;- wide_founder_sample_genos %&gt;%
  dplyr::select(sample) %&gt;%
  dplyr::left_join(.,final_founder_samples %&gt;%
                     dplyr::select(sample, dam, sire) %&gt;%
                     dplyr::mutate(dam = as.character(dam),
                                   sire = as.character(sire))) %&gt;%
  dplyr::mutate(bg = if_else(condition = dam == sire, 
                             true = paste0(dam,&quot;_&quot;,sample), 
                             false = paste0(dam,&quot;x&quot;,sire,&quot;_&quot;,sample)))
rownames(recoded_wide_sample_genos) &lt;- dend.labels$bg

# Scaling the genotype matrix, then calculating euclidean distance between all samples
dd &lt;- dist(scale(recoded_wide_sample_genos), method = &quot;euclidean&quot;)
hc &lt;- hclust(dd, method = &quot;ward.D2&quot;)

# Plotting sample distances as a dendrogram
ndendcolors &lt;- 3
dend_colors = c(brewer.pal(ndendcolors,&quot;Paired&quot;)) 
clus8 = cutree(hc, ndendcolors)
plot(as.phylo(hc), type = &quot;f&quot;, cex = 0.8, tip.color = dend_colors[clus8],
     no.margin = T, label.offset = 1, edge.width = 0.5)</code></pre>
<p><img src="figure/MUGA_Reference_QC.Rmd/founder%20sample%20dendrogram-1.png" width="960" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-no-spaces-founder-sample-dendrogram-1">
Past versions of "founder sample dendrogram-1.png"
</button>
</p>
<div id="fig-no-spaces-founder-sample-dendrogram-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/sam-widmayer/MUGA_reference_data/blob/3501596652f350afae3250facfce0abffc9ee2f5/docs/figure/MUGA_Reference_QC.Rmd/founder sample dendrogram-1.png" target="_blank">3501596</a>
</td>
<td>
sam-widmayer
</td>
<td>
2022-11-09
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>From here we curated a set of genotypes for each CC/DO founder strain
that were fixed across replicate samples from that strain.</p>
<pre class="r"><code>for(f in strict_founder_strains){
  
  print(paste(&quot;Generating Calls for&quot;,f))
  
  # Pulling the samples and genotypes for each CC/DO founder strain
  founder.geno.array &lt;- final_founder_samples %&gt;%
    dplyr::filter(dam == f,
                  sire == f) %&gt;%
    # Attach all genotypes
    dplyr::left_join(.,genos.flagged, by = &quot;sample&quot;) %&gt;%
    # Use only high-quality markers
    dplyr::filter(marker_flag != &quot;FLAG&quot;)


  # Count the number of unique allele calls for each marker
  founder.allele.counts &lt;- founder.geno.array %&gt;%
    dplyr::group_by(marker, genotype) %&gt;%
    dplyr::count()
  
  # Collect markers which have identical genotypes across samples from the same founder
  complete_founder_genos &lt;- founder.allele.counts %&gt;%
    dplyr::filter(n == max(founder.allele.counts$n)) %&gt;%
    dplyr::select(-n) %&gt;%
    `colnames&lt;-`(c(&quot;marker&quot;,f))

  # Assign these calls to a founder object
  assign(paste0(&quot;Calls_&quot;,f), complete_founder_genos)

}</code></pre>
<pre><code>[1] &quot;Generating Calls for 129S1/SvImJ&quot;
[1] &quot;Generating Calls for A/J&quot;
[1] &quot;Generating Calls for C57BL/6J&quot;
[1] &quot;Generating Calls for CAST/EiJ&quot;
[1] &quot;Generating Calls for NOD/ShiLtJ&quot;
[1] &quot;Generating Calls for NZO/HlLtJ&quot;
[1] &quot;Generating Calls for PWK/PhJ&quot;
[1] &quot;Generating Calls for WSB/EiJ&quot;</code></pre>
<p>The genotypes for intersecting markers between two strains were
combined (“crossed”) to form predicted genotypes for each F1 hybrid of
CC/DO founders. Then, the genotypes of each CC/DO founder F1 hybrid were
compared directly to what was predicted, and the concordance shown below
is the proportion of markers of each individual that match this
prediction.</p>
<pre class="r"><code># Build a list of founder consensus genotypes for good markers
founder_consensus_calls &lt;- list(`Calls_A/J`, `Calls_C57BL/6J`, `Calls_129S1/SvImJ`, `Calls_NOD/ShiLtJ`,
                                `Calls_NZO/HlLtJ`, `Calls_CAST/EiJ`, `Calls_PWK/PhJ`, `Calls_WSB/EiJ`)

# Generate a data frame with good markers as the sole column
good_markers &lt;- genos.flagged %&gt;%
  dplyr::distinct(marker, marker_flag) %&gt;%
  dplyr::filter(marker_flag != &quot;FLAG&quot;)

# Loop through the exisitng consensus calls and filter down to good markers
filtered_consensus_calls &lt;- purrr::map(founder_consensus_calls, 
                                       function(x){
                                         x %&gt;% 
                                           dplyr::filter(marker %in% good_markers$marker)
                                         }) %&gt;%
  Reduce(full_join, .)

# Filter samples down to inbred strains and attach a proper strain name column
founder_sample_metadata &lt;- final_founder_samples %&gt;%
  dplyr::left_join(., sample.predicted.sexes.df) %&gt;%
  dplyr::mutate(dam = as.character(dam),
                sire = as.character(sire),
                strain = if_else(dam == sire, true = dam, false = &quot;CROSS&quot;),
                resexed = if_else(predicted.sex != inferred.sex, true = &quot;TRUE&quot;, false = &quot;FALSE&quot;)) %&gt;% 
  dplyr::select(-inferred.sex) %&gt;%
  dplyr::filter(bg == &quot;INBRED&quot;, 
                strain %in% strict_founder_strains)
#                 sample         dam        sire predicted.sex founder     bg      strain resexed
# 1:             A/J (f)         A/J         A/J             f FOUNDER INBRED         A/J   FALSE
# 2:         A/Jm111 (m)         A/J         A/J             m FOUNDER INBRED         A/J   FALSE
# 3:        C57BL/6J (f)    C57BL/6J    C57BL/6J             f FOUNDER INBRED    C57BL/6J    TRUE
# 4:     129S1/SvImJ (f) 129S1/SvImJ 129S1/SvImJ             f FOUNDER INBRED 129S1/SvImJ   FALSE
# 5: 129S1/SvImJm212 (m) 129S1/SvImJ 129S1/SvImJ             m FOUNDER INBRED 129S1/SvImJ   FALSE
# 6:         NOD/LtJ (f)  NOD/ShiLtJ  NOD/ShiLtJ             f FOUNDER INBRED  NOD/ShiLtJ   FALSE


# Create column index for intensity data tables
founder_samples_for_ints &lt;- c(&quot;marker&quot;,founder_sample_metadata$sample)
founder_intensities &lt;- founder_sample_metadata %&gt;%
  dplyr::left_join(., long_intensities)

# 
# # Identify consensus calls that have complete data across founders
# founder_consensus_complete &lt;- filtered_consensus_calls[complete.cases(filtered_consensus_calls),]
# # Identify consensus calls that have DO NOT have complete data across founders (i.e. maybe some sort of discrepancy among samples contributing to that founder)
# founder_consensus_incomplete &lt;- filtered_consensus_calls[!complete.cases(filtered_consensus_calls),]</code></pre>
<pre class="r"><code># Assemble a list of predicted genotypes for F1s by combining each founder-specific dataframe of calls

# for(f in strict_founder_strains){
#   clean_founder_con &lt;- clean_founder_consensus_genotypes %&gt;%
#     dplyr::ungroup() %&gt;%
#     dplyr::select(marker, f)
#   assign(paste0(&quot;Clean_Calls_&quot;,f), clean_founder_con)
# }


# First form a list of dams that comprise each F1 cross type
dams &lt;- data.frame(tidyr::expand_grid(strict_founder_strains, strict_founder_strains, .name_repair = &quot;minimal&quot;)) %&gt;%
  `colnames&lt;-`(c(&quot;dams&quot;,&quot;sires&quot;)) %&gt;%
  # Select crosses between strains
  # filter(dams != sires) %&gt;%
  dplyr::select(dams) %&gt;%
  as.list()
# Loop through the list of cross types and pull in genotype call objects
dam_calls &lt;- list()
for(i in 1:length(dams$dams)){
  dam_calls[[i]] &lt;- get(ls(pattern = paste0(&quot;Calls_&quot;,dams$dams[i])))
}

# Do the same for the sires
sires &lt;- data.frame(tidyr::expand_grid(strict_founder_strains, strict_founder_strains, .name_repair = &quot;minimal&quot;)) %&gt;%
  `colnames&lt;-`(c(&quot;dams&quot;,&quot;sires&quot;)) %&gt;%
  # filter(dams != sires) %&gt;%
  dplyr::select(sires) %&gt;%
  as.list()
sire_calls &lt;- list()
for(i in 1:length(sires$sires)){
  sire_calls[[i]] &lt;- get(ls(pattern = paste0(&quot;Calls_&quot;,sires$sires[i])))
}

# Compare the predicted genotypes (from consensus calls) to the actual genotypes of each sample
# founder_background_QC(dam = dam_calls[[2]], sire = sire_calls[[2]])
bg_QC &lt;- purrr::map2(dam_calls, sire_calls, founder_background_QC)</code></pre>
<pre><code>[1] &quot;Running QC: 129S1/SvImJ&quot;
[1] &quot;Running QC: (129S1/SvImJxA/J)F1&quot;
[1] &quot;Running QC: (129S1/SvImJxC57BL/6J)F1&quot;
[1] &quot;Running QC: (129S1/SvImJxCAST/EiJ)F1&quot;
[1] &quot;Running QC: (129S1/SvImJxNOD/ShiLtJ)F1&quot;
[1] &quot;Running QC: (129S1/SvImJxNZO/HlLtJ)F1&quot;
[1] &quot;Running QC: (129S1/SvImJxPWK/PhJ)F1&quot;
[1] &quot;Running QC: (129S1/SvImJxWSB/EiJ)F1&quot;
[1] &quot;Running QC: (A/Jx129S1/SvImJ)F1&quot;
[1] &quot;Running QC: A/J&quot;
[1] &quot;Running QC: (A/JxC57BL/6J)F1&quot;
[1] &quot;Running QC: (A/JxCAST/EiJ)F1&quot;
[1] &quot;Running QC: (A/JxNOD/ShiLtJ)F1&quot;
[1] &quot;Running QC: (A/JxNZO/HlLtJ)F1&quot;
[1] &quot;Running QC: (A/JxPWK/PhJ)F1&quot;
[1] &quot;Running QC: (A/JxWSB/EiJ)F1&quot;
[1] &quot;Running QC: (C57BL/6Jx129S1/SvImJ)F1&quot;
[1] &quot;Running QC: (C57BL/6JxA/J)F1&quot;
[1] &quot;Running QC: C57BL/6J&quot;
[1] &quot;Running QC: (C57BL/6JxCAST/EiJ)F1&quot;
[1] &quot;Running QC: (C57BL/6JxNOD/ShiLtJ)F1&quot;
[1] &quot;Running QC: (C57BL/6JxNZO/HlLtJ)F1&quot;
[1] &quot;Running QC: (C57BL/6JxPWK/PhJ)F1&quot;
[1] &quot;Running QC: (C57BL/6JxWSB/EiJ)F1&quot;
[1] &quot;Running QC: (CAST/EiJx129S1/SvImJ)F1&quot;
[1] &quot;Running QC: (CAST/EiJxA/J)F1&quot;
[1] &quot;Running QC: (CAST/EiJxC57BL/6J)F1&quot;
[1] &quot;Running QC: CAST/EiJ&quot;
[1] &quot;Running QC: (CAST/EiJxNOD/ShiLtJ)F1&quot;
[1] &quot;Running QC: (CAST/EiJxNZO/HlLtJ)F1&quot;
[1] &quot;Running QC: (CAST/EiJxPWK/PhJ)F1&quot;
[1] &quot;Running QC: (CAST/EiJxWSB/EiJ)F1&quot;
[1] &quot;Running QC: (NOD/ShiLtJx129S1/SvImJ)F1&quot;
[1] &quot;Running QC: (NOD/ShiLtJxA/J)F1&quot;
[1] &quot;Running QC: (NOD/ShiLtJxC57BL/6J)F1&quot;
[1] &quot;Running QC: (NOD/ShiLtJxCAST/EiJ)F1&quot;
[1] &quot;Running QC: NOD/ShiLtJ&quot;
[1] &quot;Running QC: (NOD/ShiLtJxNZO/HlLtJ)F1&quot;
[1] &quot;Running QC: (NOD/ShiLtJxPWK/PhJ)F1&quot;
[1] &quot;Running QC: (NOD/ShiLtJxWSB/EiJ)F1&quot;
[1] &quot;Running QC: (NZO/HlLtJx129S1/SvImJ)F1&quot;
[1] &quot;Running QC: (NZO/HlLtJxA/J)F1&quot;
[1] &quot;Running QC: (NZO/HlLtJxC57BL/6J)F1&quot;
[1] &quot;Running QC: (NZO/HlLtJxCAST/EiJ)F1&quot;
[1] &quot;Running QC: (NZO/HlLtJxNOD/ShiLtJ)F1&quot;
[1] &quot;Running QC: NZO/HlLtJ&quot;
[1] &quot;Running QC: (NZO/HlLtJxPWK/PhJ)F1&quot;
[1] &quot;Running QC: (NZO/HlLtJxWSB/EiJ)F1&quot;
[1] &quot;Running QC: (PWK/PhJx129S1/SvImJ)F1&quot;
[1] &quot;Running QC: (PWK/PhJxA/J)F1&quot;
[1] &quot;Running QC: (PWK/PhJxC57BL/6J)F1&quot;
[1] &quot;Running QC: (PWK/PhJxCAST/EiJ)F1&quot;
[1] &quot;Running QC: (PWK/PhJxNOD/ShiLtJ)F1&quot;
[1] &quot;Running QC: (PWK/PhJxNZO/HlLtJ)F1&quot;
[1] &quot;Running QC: PWK/PhJ&quot;
[1] &quot;Running QC: (PWK/PhJxWSB/EiJ)F1&quot;
[1] &quot;Running QC: (WSB/EiJx129S1/SvImJ)F1&quot;
[1] &quot;Running QC: (WSB/EiJxA/J)F1&quot;
[1] &quot;Running QC: (WSB/EiJxC57BL/6J)F1&quot;
[1] &quot;Running QC: (WSB/EiJxCAST/EiJ)F1&quot;
[1] &quot;Running QC: (WSB/EiJxNOD/ShiLtJ)F1&quot;
[1] &quot;Running QC: (WSB/EiJxNZO/HlLtJ)F1&quot;
[1] &quot;Running QC: (WSB/EiJxPWK/PhJ)F1&quot;
[1] &quot;Running QC: WSB/EiJ&quot;</code></pre>
<pre class="r"><code># Keep outputs from the QC that are lists; if QC wasn&#39;t performed for a given background, the output was a character vector warning
founder_background_QC_tr &lt;- bg_QC %&gt;%
  purrr::keep(., is.list) %&gt;%
  # Instead of having 64 elements of lists of two, have two lists of 64:
  # 1) All good genotypes from each cross with concordance values
  # 2) All concordance summaries for each cross
  purrr::transpose(.)

# Bind together all concordance summaries
founder_concordance_df &lt;- Reduce(rbind, founder_background_QC_tr[[2]])
# If all markers for a given chromosome type were either concordant or discordant, NAs are returned
# This step assigns those NA values a 0
founder_concordance_df[is.na(founder_concordance_df)] &lt;- 0
# Form concordance as a percentage
founder_concordance_df_2 &lt;- founder_concordance_df %&gt;%
  dplyr::mutate(concordance = MATCH/(MATCH + `NO MATCH`)) %&gt;%
  dplyr::mutate(dam = gsub(dam, pattern = &quot;[.]&quot;, replacement = &quot;/&quot;),
                sire = gsub(sire, pattern = &quot;[.]&quot;, replacement = &quot;/&quot;))
founder_concordance_df_2$alt_chr &lt;- factor(founder_concordance_df_2$alt_chr,
                                           levels = c(&quot;Autosome&quot;,&quot;X&quot;,&quot;Y&quot;,&quot;M&quot;,&quot;Other&quot;))
founder_concordance_df_3 &lt;- founder_concordance_df_2 %&gt;%
  dplyr::filter(alt_chr == &quot;Autosome&quot;)
founder_palette &lt;- qtl2::CCcolors
names(founder_palette) &lt;- strict_founder_strains[c(2,3,1,5,6,4,7,8)]
founder_palette_2 &lt;- founder_palette

concordance_plot &lt;- ggplot(founder_concordance_df_3, mapping = aes(x = dam, y = concordance, color = dam, fill = sire)) + 
  geom_jitter(shape = 21, width = 0.25) +
  scale_fill_manual(values = founder_palette_2) +
  scale_colour_manual(values = founder_palette_2) +
  ylim(c(0.5,1.05)) + 
  labs(x = &quot;Dam Founder Strain&quot;,
       y = &quot;Autosomal Genotype Concordance&quot;) + 
  QCtheme
ggplotly(concordance_plot)</code></pre>
<div id="htmlwidget-6d80764996365cf99cc5" style="width:960px;height:480px;" class="plotly html-widget"></div>
<script type="application/json" data-for="htmlwidget-6d80764996365cf99cc5">{"x":{"data":[{"x":[1.13469619804528,1.1208039522171],"y":[0.999999727521387,0.999999884100239],"text":"dam: 129S1/SvImJ<br />concordance: 1.0000000<br />dam: 129S1/SvImJ<br />sire: 129S1/SvImJ","type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(230,159,0,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(230,159,0,1)"}},"hoveron":"points","name":"129S1/SvImJ","legendgroup":"129S1/SvImJ","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[4.02163078065496],"y":[0.996842955706636],"text":"dam: CAST/EiJ<br />concordance: 0.9968428<br />dam: CAST/EiJ<br />sire: 129S1/SvImJ","type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(230,159,0,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,158,115,1)"}},"hoveron":"points","name":"(129S1/SvImJ,CAST/EiJ)","legendgroup":"(129S1/SvImJ,CAST/EiJ)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[1.2178737615468],"y":[0.996730588703866],"text":"dam: 129S1/SvImJ<br />concordance: 0.9967304<br />dam: 129S1/SvImJ<br />sire: A/J","type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(240,228,66,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(230,159,0,1)"}},"hoveron":"points","name":"(A/J,129S1/SvImJ)","legendgroup":"(A/J,129S1/SvImJ)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[1.75687108270358,2.01310374366585],"y":[0.999999925512101,1.0000001367099],"text":"dam: A/J<br />concordance: 1.0000000<br />dam: A/J<br />sire: A/J","type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(240,228,66,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(240,228,66,1)"}},"hoveron":"points","name":"A/J","legendgroup":"A/J","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[4.06658231909387,4.10364392807242],"y":[0.996689145201012,0.996355400655125],"text":["dam: CAST/EiJ<br />concordance: 0.9966893<br />dam: CAST/EiJ<br />sire: A/J","dam: CAST/EiJ<br />concordance: 0.9963552<br />dam: CAST/EiJ<br />sire: A/J"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(240,228,66,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,158,115,1)"}},"hoveron":"points","name":"(A/J,CAST/EiJ)","legendgroup":"(A/J,CAST/EiJ)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[5.08559825725388],"y":[0.997200567696367],"text":"dam: NOD/ShiLtJ<br />concordance: 0.9972007<br />dam: NOD/ShiLtJ<br />sire: A/J","type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(240,228,66,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,114,178,1)"}},"hoveron":"points","name":"(A/J,NOD/ShiLtJ)","legendgroup":"(A/J,NOD/ShiLtJ)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[0.751418676925823],"y":[0.992083157369894],"text":"dam: 129S1/SvImJ<br />concordance: 0.9920831<br />dam: 129S1/SvImJ<br />sire: C57BL/6J","type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(85,85,85,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(230,159,0,1)"}},"hoveron":"points","name":"(C57BL/6J,129S1/SvImJ)","legendgroup":"(C57BL/6J,129S1/SvImJ)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[2.23459153459407],"y":[0.995399840526913],"text":"dam: A/J<br />concordance: 0.9954000<br />dam: A/J<br />sire: C57BL/6J","type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(85,85,85,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(240,228,66,1)"}},"hoveron":"points","name":"(C57BL/6J,A/J)","legendgroup":"(C57BL/6J,A/J)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[2.9210316388635],"y":[0.999999920245611],"text":"dam: C57BL/6J<br />concordance: 1.0000000<br />dam: C57BL/6J<br />sire: C57BL/6J","type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(85,85,85,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(85,85,85,1)"}},"hoveron":"points","name":"C57BL/6J","legendgroup":"C57BL/6J","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[0.950871231383644],"y":[0.996845924599435],"text":"dam: 129S1/SvImJ<br />concordance: 0.9968459<br />dam: 129S1/SvImJ<br />sire: CAST/EiJ","type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(0,158,115,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(230,159,0,1)"}},"hoveron":"points","name":"(CAST/EiJ,129S1/SvImJ)","legendgroup":"(CAST/EiJ,129S1/SvImJ)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[2.83391502557788],"y":[0.992125992506429],"text":"dam: C57BL/6J<br />concordance: 0.9921260<br />dam: C57BL/6J<br />sire: CAST/EiJ","type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(0,158,115,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(85,85,85,1)"}},"hoveron":"points","name":"(CAST/EiJ,C57BL/6J)","legendgroup":"(CAST/EiJ,C57BL/6J)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[3.84229223965667,3.92218728724401],"y":[0.999999895983683,0.999999735318452],"text":"dam: CAST/EiJ<br />concordance: 1.0000000<br />dam: CAST/EiJ<br />sire: CAST/EiJ","type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(0,158,115,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,158,115,1)"}},"hoveron":"points","name":"CAST/EiJ","legendgroup":"CAST/EiJ","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[8.01760540285613],"y":[0.99834054210069],"text":"dam: WSB/EiJ<br />concordance: 0.9983405<br />dam: WSB/EiJ<br />sire: CAST/EiJ","type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(0,158,115,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(204,121,167,1)"}},"hoveron":"points","name":"(CAST/EiJ,WSB/EiJ)","legendgroup":"(CAST/EiJ,WSB/EiJ)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[0.97840205708053],"y":[0.995854592224636],"text":"dam: 129S1/SvImJ<br />concordance: 0.9958548<br />dam: 129S1/SvImJ<br />sire: NOD/ShiLtJ","type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(0,114,178,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(230,159,0,1)"}},"hoveron":"points","name":"(NOD/ShiLtJ,129S1/SvImJ)","legendgroup":"(NOD/ShiLtJ,129S1/SvImJ)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[3.07778558880091],"y":[0.995676650968003],"text":"dam: C57BL/6J<br />concordance: 0.9956768<br />dam: C57BL/6J<br />sire: NOD/ShiLtJ","type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(0,114,178,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(85,85,85,1)"}},"hoveron":"points","name":"(NOD/ShiLtJ,C57BL/6J)","legendgroup":"(NOD/ShiLtJ,C57BL/6J)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[3.80086158332415],"y":[0.996806370635452],"text":"dam: CAST/EiJ<br />concordance: 0.9968062<br />dam: CAST/EiJ<br />sire: NOD/ShiLtJ","type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(0,114,178,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,158,115,1)"}},"hoveron":"points","name":"(NOD/ShiLtJ,CAST/EiJ)","legendgroup":"(NOD/ShiLtJ,CAST/EiJ)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[5.01176029955968],"y":[1.0000000859704],"text":"dam: NOD/ShiLtJ<br />concordance: 1.0000000<br />dam: NOD/ShiLtJ<br />sire: NOD/ShiLtJ","type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(0,114,178,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,114,178,1)"}},"hoveron":"points","name":"NOD/ShiLtJ","legendgroup":"NOD/ShiLtJ","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[1.23574898694642],"y":[0.995071533214733],"text":"dam: 129S1/SvImJ<br />concordance: 0.9950715<br />dam: 129S1/SvImJ<br />sire: NZO/HlLtJ","type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(86,180,233,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(230,159,0,1)"}},"hoveron":"points","name":"(NZO/HlLtJ,129S1/SvImJ)","legendgroup":"(NZO/HlLtJ,129S1/SvImJ)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[1.76614064082969],"y":[0.99755693867418],"text":"dam: A/J<br />concordance: 0.9975570<br />dam: A/J<br />sire: NZO/HlLtJ","type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(86,180,233,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(240,228,66,1)"}},"hoveron":"points","name":"(NZO/HlLtJ,A/J)","legendgroup":"(NZO/HlLtJ,A/J)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[2.94410098367371],"y":[0.993763162225594],"text":"dam: C57BL/6J<br />concordance: 0.9937633<br />dam: C57BL/6J<br />sire: NZO/HlLtJ","type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(86,180,233,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(85,85,85,1)"}},"hoveron":"points","name":"(NZO/HlLtJ,C57BL/6J)","legendgroup":"(NZO/HlLtJ,C57BL/6J)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[5.13880512444302],"y":[0.997014508581259],"text":"dam: NOD/ShiLtJ<br />concordance: 0.9970144<br />dam: NOD/ShiLtJ<br />sire: NZO/HlLtJ","type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(86,180,233,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,114,178,1)"}},"hoveron":"points","name":"(NZO/HlLtJ,NOD/ShiLtJ)","legendgroup":"(NZO/HlLtJ,NOD/ShiLtJ)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[6.14669589174446,5.88817063905299],"y":[0.999999931737247,1.00000025991187],"text":"dam: NZO/HlLtJ<br />concordance: 1.0000000<br />dam: NZO/HlLtJ<br />sire: NZO/HlLtJ","type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(86,180,233,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(86,180,233,1)"}},"hoveron":"points","name":"NZO/HlLtJ","legendgroup":"NZO/HlLtJ","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[1.19027432752773,1.17727218766231],"y":[0.995991804385695,0.996158333568365],"text":["dam: 129S1/SvImJ<br />concordance: 0.9959920<br />dam: 129S1/SvImJ<br />sire: PWK/PhJ","dam: 129S1/SvImJ<br />concordance: 0.9961583<br />dam: 129S1/SvImJ<br />sire: PWK/PhJ"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(213,94,0,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(230,159,0,1)"}},"hoveron":"points","name":"(PWK/PhJ,129S1/SvImJ)","legendgroup":"(PWK/PhJ,129S1/SvImJ)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[1.87949015956838,2.12008482194506],"y":[0.99533874792144,0.994839315746228],"text":["dam: A/J<br />concordance: 0.9953388<br />dam: A/J<br />sire: PWK/PhJ","dam: A/J<br />concordance: 0.9948394<br />dam: A/J<br />sire: PWK/PhJ"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(213,94,0,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(240,228,66,1)"}},"hoveron":"points","name":"(PWK/PhJ,A/J)","legendgroup":"(PWK/PhJ,A/J)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[3.23653678630944,2.76365774928126],"y":[0.994196724893903,0.995192485831797],"text":["dam: C57BL/6J<br />concordance: 0.9941967<br />dam: C57BL/6J<br />sire: PWK/PhJ","dam: C57BL/6J<br />concordance: 0.9951923<br />dam: C57BL/6J<br />sire: PWK/PhJ"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(213,94,0,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(85,85,85,1)"}},"hoveron":"points","name":"(PWK/PhJ,C57BL/6J)","legendgroup":"(PWK/PhJ,C57BL/6J)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[3.85058152419515],"y":[0.997126461792323],"text":"dam: CAST/EiJ<br />concordance: 0.9971264<br />dam: CAST/EiJ<br />sire: PWK/PhJ","type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(213,94,0,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,158,115,1)"}},"hoveron":"points","name":"(PWK/PhJ,CAST/EiJ)","legendgroup":"(PWK/PhJ,CAST/EiJ)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[4.85318831761833,4.75498418766074],"y":[0.995950100662324,0.995949150007778],"text":["dam: NOD/ShiLtJ<br />concordance: 0.9959501<br />dam: NOD/ShiLtJ<br />sire: PWK/PhJ","dam: NOD/ShiLtJ<br />concordance: 0.9959494<br />dam: NOD/ShiLtJ<br />sire: PWK/PhJ"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(213,94,0,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,114,178,1)"}},"hoveron":"points","name":"(PWK/PhJ,NOD/ShiLtJ)","legendgroup":"(PWK/PhJ,NOD/ShiLtJ)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[6.79606315982528,7.01390417292714],"y":[0.999999998698761,1.00000026658393],"text":"dam: PWK/PhJ<br />concordance: 1.0000000<br />dam: PWK/PhJ<br />sire: PWK/PhJ","type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(213,94,0,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(213,94,0,1)"}},"hoveron":"points","name":"PWK/PhJ","legendgroup":"PWK/PhJ","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[8.23543115344364,8.05048716219608],"y":[0.9976615826919,0.997661500700401],"text":"dam: WSB/EiJ<br />concordance: 0.9976616<br />dam: WSB/EiJ<br />sire: PWK/PhJ","type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(213,94,0,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(204,121,167,1)"}},"hoveron":"points","name":"(PWK/PhJ,WSB/EiJ)","legendgroup":"(PWK/PhJ,WSB/EiJ)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[1.1967917609727],"y":[0.996362925021744],"text":"dam: 129S1/SvImJ<br />concordance: 0.9963630<br />dam: 129S1/SvImJ<br />sire: WSB/EiJ","type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(204,121,167,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(230,159,0,1)"}},"hoveron":"points","name":"(WSB/EiJ,129S1/SvImJ)","legendgroup":"(WSB/EiJ,129S1/SvImJ)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[1.92672260047402],"y":[0.997365759890727],"text":"dam: A/J<br />concordance: 0.9973658<br />dam: A/J<br />sire: WSB/EiJ","type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(204,121,167,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(240,228,66,1)"}},"hoveron":"points","name":"(WSB/EiJ,A/J)","legendgroup":"(WSB/EiJ,A/J)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[2.83116680767853],"y":[0.995623137353846],"text":"dam: C57BL/6J<br />concordance: 0.9956233<br />dam: C57BL/6J<br />sire: WSB/EiJ","type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(204,121,167,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(85,85,85,1)"}},"hoveron":"points","name":"(WSB/EiJ,C57BL/6J)","legendgroup":"(WSB/EiJ,C57BL/6J)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[4.0651480598608,3.88843102729879],"y":[0.998009509274233,0.997842442012368],"text":["dam: CAST/EiJ<br />concordance: 0.9980093<br />dam: CAST/EiJ<br />sire: WSB/EiJ","dam: CAST/EiJ<br />concordance: 0.9978427<br />dam: CAST/EiJ<br />sire: WSB/EiJ"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(204,121,167,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,158,115,1)"}},"hoveron":"points","name":"(WSB/EiJ,CAST/EiJ)","legendgroup":"(WSB/EiJ,CAST/EiJ)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[5.11028015380725],"y":[0.997839527084875],"text":"dam: NOD/ShiLtJ<br />concordance: 0.9978395<br />dam: NOD/ShiLtJ<br />sire: WSB/EiJ","type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(204,121,167,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,114,178,1)"}},"hoveron":"points","name":"(WSB/EiJ,NOD/ShiLtJ)","legendgroup":"(WSB/EiJ,NOD/ShiLtJ)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[6.02952500327956],"y":[0.996367438549663],"text":"dam: NZO/HlLtJ<br />concordance: 0.9963672<br />dam: NZO/HlLtJ<br />sire: WSB/EiJ","type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(204,121,167,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(86,180,233,1)"}},"hoveron":"points","name":"(WSB/EiJ,NZO/HlLtJ)","legendgroup":"(WSB/EiJ,NZO/HlLtJ)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[8.24757337360643],"y":[0.999999746148857],"text":"dam: WSB/EiJ<br />concordance: 1.0000000<br />dam: WSB/EiJ<br />sire: WSB/EiJ","type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(204,121,167,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(204,121,167,1)"}},"hoveron":"points","name":"WSB/EiJ","legendgroup":"WSB/EiJ","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null}],"layout":{"margin":{"t":30.6118721461187,"r":7.30593607305936,"b":44.5662100456621,"l":43.1050228310502},"plot_bgcolor":"rgba(255,255,255,1)","paper_bgcolor":"rgba(255,255,255,1)","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187},"xaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[0.4,8.6],"tickmode":"array","ticktext":["129S1/SvImJ","A/J","C57BL/6J","CAST/EiJ","NOD/ShiLtJ","NZO/HlLtJ","PWK/PhJ","WSB/EiJ"],"tickvals":[1,2,3,4,5,6,7,8],"categoryorder":"array","categoryarray":["129S1/SvImJ","A/J","C57BL/6J","CAST/EiJ","NOD/ShiLtJ","NZO/HlLtJ","PWK/PhJ","WSB/EiJ"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.65296803652968,"tickwidth":0.66417600664176,"showticklabels":true,"tickfont":{"color":"rgba(0,0,0,1)","family":"","size":11.689497716895},"tickangle":-0,"showline":false,"linecolor":null,"linewidth":0,"showgrid":false,"gridcolor":null,"gridwidth":0,"zeroline":false,"anchor":"y","title":{"text":"Dam Founder Strain","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187}},"hoverformat":".2f"},"yaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[0.4725,1.0775],"tickmode":"array","ticktext":["0.5","0.6","0.7","0.8","0.9","1.0"],"tickvals":[0.5,0.6,0.7,0.8,0.9,1],"categoryorder":"array","categoryarray":["0.5","0.6","0.7","0.8","0.9","1.0"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.65296803652968,"tickwidth":0.66417600664176,"showticklabels":true,"tickfont":{"color":"rgba(0,0,0,1)","family":"","size":11.689497716895},"tickangle":-0,"showline":false,"linecolor":null,"linewidth":0,"showgrid":false,"gridcolor":null,"gridwidth":0,"zeroline":false,"anchor":"x","title":{"text":"Autosomal Genotype Concordance","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187}},"hoverformat":".2f"},"shapes":[{"type":"rect","fillcolor":"transparent","line":{"color":"rgba(51,51,51,1)","width":0.66417600664176,"linetype":"solid"},"yref":"paper","xref":"paper","x0":0,"x1":1,"y0":0,"y1":1}],"showlegend":true,"legend":{"bgcolor":"rgba(255,255,255,1)","bordercolor":"transparent","borderwidth":1.88976377952756,"font":{"color":"rgba(0,0,0,1)","family":"","size":11.689497716895},"title":{"text":"dam<br />sire","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187}}},"hovermode":"closest","barmode":"relative"},"config":{"doubleClick":"reset","modeBarButtonsToAdd":["hoverclosest","hovercompare"],"showSendToCloud":false},"source":"A","attrs":{"80cc1f906b83":{"x":{},"y":{},"colour":{},"fill":{},"type":"scatter"}},"cur_data":"80cc1f906b83","visdat":{"80cc1f906b83":["function (y) ","x"]},"highlight":{"on":"plotly_click","persistent":false,"dynamic":false,"selectize":false,"opacityDim":0.2,"selected":{"opacity":1},"debounce":0},"shinyEvents":["plotly_hover","plotly_click","plotly_selected","plotly_relayout","plotly_brushed","plotly_brushing","plotly_clickannotation","plotly_doubleclick","plotly_deselect","plotly_afterplot","plotly_sunburstclick"],"base_url":"https://plot.ly"},"evals":[],"jsHooks":[]}</script>
</div>
</div>
<div id="writing-reference-files" class="section level2">
<h2>Writing reference files</h2>
<p>The list of output file types following QC is as follows:</p>
<ul>
<li><p>Genotype file; rows = markers, columns = samples</p></li>
<li><p>Probe intensity file; rows = markers, columns = samples</p></li>
<li><p>Sample metadata; columns = sample, strain, sex</p></li>
</ul>
<p>Each file type is generated for:</p>
<ul>
<li><p>All good samples</p></li>
<li><p>Eight CC/DO founder strains</p></li>
</ul>
<p>In total, 7762 markers and 53 samples passed the QC steps we imposed,
and wrote these genotype data and metadata to new reference files.</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.2.1 (2022-06-23)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Big Sur ... 10.16

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] vroom_1.6.0        tictoc_1.1         RColorBrewer_1.1-3 ape_5.6-2         
 [5] progress_1.2.2     plotly_4.10.0      ggplot2_3.3.6      DT_0.25           
 [9] magrittr_2.0.3     qtl2_0.28          furrr_0.3.1        future_1.28.0     
[13] purrr_0.3.4        data.table_1.14.2  stringr_1.4.1      tidyr_1.2.1       
[17] dplyr_1.0.10       workflowr_1.7.0   

loaded via a namespace (and not attached):
 [1] httr_1.4.4        sass_0.4.2        bit64_4.0.5       jsonlite_1.8.0   
 [5] viridisLite_0.4.1 bslib_0.4.0       assertthat_0.2.1  getPass_0.2-2    
 [9] highr_0.9         blob_1.2.3        yaml_2.3.5        globals_0.16.1   
[13] pillar_1.8.1      RSQLite_2.2.17    lattice_0.20-45   glue_1.6.2       
[17] digest_0.6.29     promises_1.2.0.1  colorspace_2.0-3  htmltools_0.5.3  
[21] httpuv_1.6.6      pkgconfig_2.0.3   listenv_0.8.0     scales_1.2.1     
[25] processx_3.7.0    whisker_0.4       later_1.3.0       tzdb_0.3.0       
[29] git2r_0.30.1      tibble_3.1.8      farver_2.1.1      generics_0.1.3   
[33] ellipsis_0.3.2    cachem_1.0.6      withr_2.5.0       lazyeval_0.2.2   
[37] cli_3.4.0         crayon_1.5.1      memoise_2.0.1     evaluate_0.16    
[41] ps_1.7.1          fs_1.5.2          fansi_1.0.3       parallelly_1.32.1
[45] nlme_3.1-157      tools_4.2.1       prettyunits_1.1.1 hms_1.1.2        
[49] lifecycle_1.0.2   munsell_0.5.0     callr_3.7.2       compiler_4.2.1   
[53] jquerylib_0.1.4   rlang_1.0.5       grid_4.2.1        rstudioapi_0.14  
[57] htmlwidgets_1.5.4 crosstalk_1.2.0   labeling_0.4.2    rmarkdown_2.16   
[61] gtable_0.3.1      codetools_0.2-18  DBI_1.1.3         R6_2.5.1         
[65] knitr_1.40        fastmap_1.1.0     bit_4.0.4         utf8_1.2.2       
[69] rprojroot_2.0.3   stringi_1.7.8     parallel_4.2.1    Rcpp_1.0.9       
[73] vctrs_0.4.1       tidyselect_1.1.2  xfun_0.33        </code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
